<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Grammar of Graphics for Comparative Genomics • gggenomes</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<script src="extra.js"></script><meta property="og:title" content="A Grammar of Graphics for Comparative Genomics">
<meta property="og:description" content="gggenomes is an extension of ggplot2 for creating complex genomic
 maps. It builds on the power of ggplot2 and tidyverse adding new ggplot-style
 geoms &amp; positions and dplyr-style verbs to manipulate the underlying data. It
 implements a layout concept inspired by ggraph and introduces tracks to bring
 tidyness to the mess that is genomics data.">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/emales.html">From a few sequences to a complex map in minutes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thackl/gggenomes/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="gggenomes-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#gggenomes-" class="anchor"></a>gggenomes <img src="reference/figures/logo-smaller.png" align="right">
</h1></div>
<div id="a-grammar-of-graphics-for-comparative-genomics" class="section level2">
<h2 class="hasAnchor">
<a href="#a-grammar-of-graphics-for-comparative-genomics" class="anchor"></a>A grammar of graphics for comparative genomics</h2>
<p>gggenomes is a versatile graphics package for creating comparative genomic maps, extending the popular R visualization package <a href="https://ggplot2.tidyverse.org/">ggplot2</a>. It adds dedicated plot functions (geoms, positions) for common sequence and alignment features, such as gene models and syntenic regions, as we well as verbs to further manipulate the plot and the underlying data, for example, to quickly zoom in into gene neighborhoods.</p>
</div>
<div id="a-realistic-use-case-comparing-six-viral-genomes" class="section level2">
<h2 class="hasAnchor">
<a href="#a-realistic-use-case-comparing-six-viral-genomes" class="anchor"></a>A realistic use case comparing six viral genomes</h2>
<p>For a reproducible recipe describing the full <em>evolution</em> of this plot starting from a mere set of contigs, and including the bioinformatics analysis workflow, have a look at <a href="https://thackl.github.io/gggenomes/articles/emales.html">From a few sequences to a complex map in minutes</a>. For a compact version to of the code see below.</p>
<p><img src="reference/figures/EMALEs.png"></p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="co"># EMALEs: endogenous mavirus-like elements (Hackl et al. in prep.)</span>
<span class="co"># a use case data bundled with ggggenomes</span>
<span class="no">seqs</span> <span class="kw">&lt;-</span> <span class="no">emale_seqs</span>[<span class="fl">1</span>:<span class="fl">6</span>,]                <span class="co"># first 6 genomes only</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/gggenomes.html">gggenomes</a></span>(<span class="no">seqs</span>, <span class="no">emale_genes</span>, <span class="no">emale_tirs</span>, <span class="no">emale_links</span>) <span class="kw">%&gt;%</span>
                                        <span class="co"># standard tracks</span>
  <span class="fu"><a href="reference/add_features.html">add_features</a></span>(<span class="no">emale_transposons</span>, <span class="no">emale_gc</span>) <span class="kw">%&gt;%</span>
                                        <span class="co"># more features</span>
  <span class="fu"><a href="reference/add_subfeatures.html">add_subfeatures</a></span>(<span class="no">genes</span>, <span class="no">emale_blast</span>, <span class="kw">transform</span><span class="kw">=</span><span class="st">"aa2nuc"</span>) <span class="kw">%&gt;%</span>
                                        <span class="co"># features that map onto features</span>
  <span class="fu"><a href="reference/add_clusters.html">add_clusters</a></span>(<span class="no">genes</span>, <span class="no">emale_cogs</span>)       <span class="co"># and some gene cluster info</span>

<span class="co"># add geoms and customize the plot</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="no">p</span> +
  <span class="fu"><a href="reference/geom_seq.html">geom_seq</a></span>() + <span class="fu"><a href="reference/geom_bin_label.html">geom_bin_label</a></span>() +       <span class="co"># chromosomes and labels</span>
  <span class="fu"><a href="reference/geom_link.html">geom_link</a></span>(<span class="kw">offset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.3</span>, <span class="fl">0.2</span>), <span class="kw">color</span><span class="kw">=</span><span class="st">"white"</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="fl">.3</span>) +
                                        <span class="co"># synthenic regions</span>
  <span class="fu"><a href="reference/geom_feature.html">geom_feature</a></span>(<span class="fu">aes</span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"terminal inverted repeat"</span>), <span class="fu"><a href="reference/use_seqs.html">use_features</a></span>(<span class="fl">2</span>),
    <span class="kw">size</span><span class="kw">=</span><span class="fl">4</span>) +                           <span class="co"># repeats</span>
  <span class="fu"><a href="reference/geom_feature.html">geom_feature</a></span>(<span class="fu">aes</span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"integrated transposon"</span>),
    <span class="fu"><a href="reference/use_seqs.html">use_features</a></span>(<span class="no">emale_transposons</span>), <span class="kw">size</span><span class="kw">=</span><span class="fl">7</span>) +
                                        <span class="co"># transposons</span>
  <span class="fu"><a href="reference/geom_gene.html">geom_gene</a></span>(<span class="fu">aes</span>(<span class="kw">fill</span><span class="kw">=</span><span class="no">cluster_label</span>)) +  <span class="co"># colored genes</span>
  <span class="fu"><a href="reference/geom_feature.html">geom_feature</a></span>(<span class="fu">aes</span>(<span class="kw">color</span><span class="kw">=</span><span class="no">blast_desc</span>), <span class="fu"><a href="reference/use_seqs.html">use_features</a></span>(<span class="no">emale_blast</span>),
    <span class="kw">size</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">position</span><span class="kw">=</span><span class="st">"pile"</span>) +          <span class="co"># some blast hits</span>
  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="kw">x</span><span class="kw">=</span>(<span class="no">x</span>+<span class="no">xend</span>)/<span class="fl">2</span>, <span class="kw">ymax</span><span class="kw">=</span><span class="no">y</span>+<span class="fl">.24</span>, <span class="kw">ymin</span><span class="kw">=</span><span class="no">y</span>+<span class="fl">.38</span>-(<span class="fl">.4</span>*<span class="no">score</span>),
    <span class="kw">group</span><span class="kw">=</span><span class="no">seq_id</span>, <span class="kw">linetype</span><span class="kw">=</span><span class="st">"GC-content"</span>), <span class="fu"><a href="reference/use_seqs.html">use_features</a></span>(<span class="no">emale_gc</span>),
    <span class="kw">fill</span><span class="kw">=</span><span class="st">"blue"</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="fl">.5</span>) +            <span class="co"># combine with ggplot2 geoms</span>
  <span class="fu">scale_fill_brewer</span>(<span class="st">"Conserved genes"</span>, <span class="kw">palette</span><span class="kw">=</span><span class="st">"Set3"</span>) +
  <span class="fu">scale_color_viridis_d</span>(<span class="st">"Blast hits &amp; Features"</span>, <span class="kw">direction</span> <span class="kw">=</span> -<span class="fl">1</span>) +
  <span class="fu">scale_linetype</span>(<span class="st">"Graphs"</span>) +
  <span class="fu">ggtitle</span>(<span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"Endogenous mavirus-like elements of "</span>,
  <span class="fu"><a href="https://rdrr.io/r/grDevices/plotmath.html">italic</a></span>(<span class="st">"C. burkhardae"</span>))))

<span class="co"># reverse-complement 3 genomes for better alignment</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="no">p</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="reference/flip.html">flip_bins</a></span>(<span class="fl">3</span>:<span class="fl">5</span>)
<span class="no">p</span></pre></div>
</div>
<div id="motivation--concept" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation--concept" class="anchor"></a>Motivation &amp; concept</h2>
<p>Visualization is a corner stone of both exploratory analysis and science communication. Bioinformatics workflows, unfortunately, tend to generate a plethora of data products often in adventurous formats making it quite difficult to integrate and co-visualize the results. Instead of trying to cater to the all these different formats explicitly, gggenomes embraces the simple tidyverse-inspired credo:</p>
<ul>
<li>Any data set can be transformed into one (or a few) tidy data tables</li>
<li>Any data set in a tidy data table can be easily and elegantly visualized</li>
</ul>
<p>As a result gggenomes helps bridge the gap between data generation, visual exploration, interpretation and communication, thereby accelerating biological research.</p>
<p>Under the hood gggenomes uses a light-weight track system to accommodate a mix of related data sets, essentially implementing ggplot2 with multiple tidy tables instead of just one. The data in the different tables are tied together through a global genome layout that is automatically computed from the input and defines the positions of genomic sequences (chromosome/contigs) and their associated features in the plot.</p>
</div>
<div id="inspiration" class="section level2">
<h2 class="hasAnchor">
<a href="#inspiration" class="anchor"></a>Inspiration</h2>
<p>gggenomes stands on the shoulder of giants. It was born out of admiration of <a href="https://wilkox.org/">David Wilkins’</a> <a href="https://github.com/wilkox/gggenes">gggenes</a> package, draws from other ggplot2 extensions such as Guangchuang Yu’s <a href="https://guangchuangyu.github.io/software/ggtree/">ggtree</a>, and is fundamentally inspired by <a href="https://www.data-imaginist.com/about">Thomas Lin Pedersen’s</a> incredibly rich <a href="https://github.com/thomasp85/ggraph">ggraph</a> package.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>gggenomes is at this point still in an alpha release state, and therefoe only available as developmental package.</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="co"># install.packages("devtools")</span>
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu">install_github</span>(<span class="st">"thackl/gggenomes"</span>)</pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/thackl/gggenomes/">https://​github.com/​thackl/​gggenomes/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/thackl/gggenomes/issues">https://​github.com/​thackl/​gggenomes/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/mit-license.php">MIT</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Thomas Hackl <br><small class="roles"> Author, maintainer </small>  </li>
<li>Markus J. Ankenbrand <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
