<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute a layout for links linking feats — as_sublinks • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Compute a layout for links linking feats — as_sublinks"><meta name="description" content="Reads sublinks connecting feats such as all-vs-all protein blasts into a
tidy dataframe. sublinks need to be associated with an already added feat
track. The sublinks are internally converted into a regular link track by
mapping the feat-based start and end coordinates to coordinates
relative to the sequences underlying the linked feats."><meta property="og:description" content="Reads sublinks connecting feats such as all-vs-all protein blasts into a
tidy dataframe. sublinks need to be associated with an already added feat
track. The sublinks are internally converted into a regular link track by
mapping the feat-based start and end coordinates to coordinates
relative to the sequences underlying the linked feats."><meta property="og:image" content="https://thackl.github.io/gggenomes/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gggenomes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/gggenomes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a></li>
    <li><a class="dropdown-item" href="../articles/flip.html">Advanced flip examples</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thackl/gggenomes/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Compute a layout for links linking feats</h1>
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/main/R/sublinks.R" class="external-link"><code>R/sublinks.R</code></a></small>
      <div class="d-none name"><code>as_sublinks.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Reads sublinks connecting feats such as all-vs-all protein blasts into a
tidy dataframe. sublinks need to be associated with an already added feat
track. The sublinks are internally converted into a regular link track by
mapping the feat-based <code>start</code> and <code>end</code> coordinates to coordinates
relative to the sequences underlying the linked feats.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">as_sublinks</span><span class="op">(</span><span class="va">x</span>, <span class="va">seqs</span>, <span class="va">feats</span>, <span class="va">...</span>, everything <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'tbl_df'</span></span>
<span><span class="fu">as_sublinks</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">seqs</span>,</span>
<span>  <span class="va">feats</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  everything <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transform <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"aa2nuc"</span>, <span class="st">"nuc2aa"</span><span class="op">)</span>,</span>
<span>  compute_layout <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>sublink data convertible to a link layout</p></dd>


<dt id="arg-seqs">seqs<a class="anchor" aria-label="anchor" href="#arg-seqs"></a></dt>
<dd><p>the sequence layout the linked feats map onto.</p></dd>


<dt id="arg-feats">feats<a class="anchor" aria-label="anchor" href="#arg-feats"></a></dt>
<dd><p>the feats the sublinks map onto.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>passed on to <code><a href="layout_seqs.html">layout_seqs()</a></code>
spaces, i.e. if matching nucleotide-level annotations to protein level
annotations, e.g. genes and protein blast results.</p></dd>


<dt id="arg-everything">everything<a class="anchor" aria-label="anchor" href="#arg-everything"></a></dt>
<dd><p>set to FALSE to drop optional columns</p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>use if feats and sublinks are in different coordinate</p></dd>


<dt id="arg-compute-layout">compute_layout<a class="anchor" aria-label="anchor" href="#arg-compute-layout"></a></dt>
<dd><p>set to FALSE to skip layout computation</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a tbl_df with plot coordinates</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The only obligatory columns are <code>feat_id</code> &amp; <code>feat_id2</code>. Also
recognized are <code>start/end</code>, <code>start2/end2</code> and <code>strand</code>.</p>
<p>Note <code>start</code> and <code>end</code> for every record will be coerced so that <code>start &lt; end</code>. If no <code>strand</code> was provided, <code>strand</code> will be added and set to "+" for
records that initially had <code>start &lt; end == start2 &lt; end2</code> and "-"
otherwise. If <code>strand</code> was provided, <code>start</code> and <code>end</code> will be reorganized to
conform with <code>start &lt; end</code> without any additional effect.</p>
    </div>
    <div class="section level2">
    <h2 id="methods-by-class-">Methods (by class)<a class="anchor" aria-label="anchor" href="#methods-by-class-"></a></h2>

<ul><li><p><code>as_sublinks(tbl_df)</code>: Convert a list of tibbles into a link layout</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

