<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Show features and regions of interest — focus • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Show features and regions of interest — focus"><meta name="description" content="Show loci containing features of interest. Loci can either be provided
as predefined regions directly (loci=), or are constructed automatically
based on pre-selected features (via ...). Features within max_dist are
greedily combined into the same locus. locate() adds these loci as new
track so that they can be easily visualized. focus() extracts those loci
from their parent sequences making them the new sequence set. These sequences
will have their locus_id as their new seq_id."><meta property="og:description" content="Show loci containing features of interest. Loci can either be provided
as predefined regions directly (loci=), or are constructed automatically
based on pre-selected features (via ...). Features within max_dist are
greedily combined into the same locus. locate() adds these loci as new
track so that they can be easily visualized. focus() extracts those loci
from their parent sequences making them the new sequence set. These sequences
will have their locus_id as their new seq_id."><meta property="og:image" content="https://thackl.github.io/gggenomes/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gggenomes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/gggenomes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a></li>
    <li><a class="dropdown-item" href="../articles/flip.html">Advanced flip examples</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thackl/gggenomes/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Show features and regions of interest</h1>
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/main/R/focus.R" class="external-link"><code>R/focus.R</code></a></small>
      <div class="d-none name"><code>focus.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Show loci containing features of interest. Loci can either be provided
as predefined regions directly (<code>loci=</code>), or are constructed automatically
based on pre-selected features (via <code>...</code>). Features within <code>max_dist</code> are
greedily combined into the same locus. <code>locate()</code> adds these loci as new
track so that they can be easily visualized. <code>focus()</code> extracts those loci
from their parent sequences making them the new sequence set. These sequences
will have their <code>locus_id</code> as their new <code>seq_id</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">focus</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  .track_id <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  .max_dist <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  .expand <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  .overhang <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"drop"</span>, <span class="st">"trim"</span>, <span class="st">"keep"</span><span class="op">)</span>,</span>
<span>  .locus_id <span class="op">=</span> <span class="fu">str_glue</span><span class="op">(</span><span class="st">"{seq_id}_lc{row_number()}"</span><span class="op">)</span>,</span>
<span>  .locus_id_group <span class="op">=</span> <span class="va">seq_id</span>,</span>
<span>  .locus_bin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bin"</span>, <span class="st">"seq"</span>, <span class="st">"locus"</span><span class="op">)</span>,</span>
<span>  .locus_score <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  .locus_filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  .loci <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">locate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  .track_id <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  .max_dist <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  .expand <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  .locus_id <span class="op">=</span> <span class="fu">str_glue</span><span class="op">(</span><span class="st">"{seq_id}_lc{row_number()}"</span><span class="op">)</span>,</span>
<span>  .locus_id_group <span class="op">=</span> <span class="va">.data</span><span class="op">$</span><span class="va">seq_id</span>,</span>
<span>  .locus_bin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bin"</span>, <span class="st">"seq"</span>, <span class="st">"locus"</span><span class="op">)</span>,</span>
<span>  .locus_score <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  .locus_filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  .locus_track <span class="op">=</span> <span class="st">"loci"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A gggenomes object</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Logical predicates defined in terms of the variables in the track
given by <code>.track_id</code>. Multiple conditions are combined with ‘&amp;’. Only rows
where the condition evaluates to ‘TRUE’ are kept.</p>
<p>The arguments in ‘...’ are automatically quoted and evaluated in the
context of the data frame. They support unquoting and splicing. See
‘vignette("programming")’ for an introduction to these concepts.</p></dd>


<dt id="arg--track-id">.track_id<a class="anchor" aria-label="anchor" href="#arg--track-id"></a></dt>
<dd><p>the track to filter from - defaults to first feature track,
usually "genes". Can be a quoted or unquoted string or a positional
argument giving the index of a track among all tracks (seqs, feats &amp;
links).</p></dd>


<dt id="arg--max-dist">.max_dist<a class="anchor" aria-label="anchor" href="#arg--max-dist"></a></dt>
<dd><p>Maximum distance between adjacent features to be included
into the same locus, default 10kb.</p></dd>


<dt id="arg--expand">.expand<a class="anchor" aria-label="anchor" href="#arg--expand"></a></dt>
<dd><p>The amount to nucleotides to expand the focus around the
target features. Default 2kb. Give two values for different up- and
downstream expansions.</p></dd>


<dt id="arg--overhang">.overhang<a class="anchor" aria-label="anchor" href="#arg--overhang"></a></dt>
<dd><p>How to handle features overlapping the locus boundaries
(including expand). Options are to "keep" them, "trim" them exactly at the
boundaries, or "drop" all features not fully included within the
boundaries.</p></dd>


<dt id="arg--locus-id-locus-id-group">.locus_id, .locus_id_group<a class="anchor" aria-label="anchor" href="#arg--locus-id-locus-id-group"></a></dt>
<dd><p>How to generate the ids for the new loci
which will eventually become their new <code>seq_id</code>s.</p></dd>


<dt id="arg--locus-bin">.locus_bin<a class="anchor" aria-label="anchor" href="#arg--locus-bin"></a></dt>
<dd><p>What bin to assign new locus to. Defaults to keeping the
original binning, but can be set to the "seq" to bin all loci originating
from the same parent sequence, or to "locus" to separate all loci into
individual bins.</p></dd>


<dt id="arg--locus-score">.locus_score<a class="anchor" aria-label="anchor" href="#arg--locus-score"></a></dt>
<dd><p>An expression evaluated in the context of all features
that are combined into a new locus. Results are stored in the column
<code>locus_score</code>. Defaults to the <code>n()</code>, i.e. the number of features per
locus. Set, for example, to <code>sum(bitscore)</code> to sum over all blast hit
bitscore of per locus. Usually used in conjunction with <code>.locus_filter</code>.</p></dd>


<dt id="arg--locus-filter">.locus_filter<a class="anchor" aria-label="anchor" href="#arg--locus-filter"></a></dt>
<dd><p>An predicate expression used to post-filter identified
loci. Set <code>.locus_filter=locus_score &gt;= 3</code> to only return loci comprising
at least 3 target features.</p></dd>


<dt id="arg--loci">.loci<a class="anchor" aria-label="anchor" href="#arg--loci"></a></dt>
<dd><p>A data.frame specifying loci directly. Required columns are
<code>seq_id,start,end</code>. Supersedes <code>...</code>.</p></dd>


<dt id="arg--locus-track">.locus_track<a class="anchor" aria-label="anchor" href="#arg--locus-track"></a></dt>
<dd><p>The name of the new track containing the identified loci.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A gggenomes object focused on the desired loci</p>
<p>A gggenomes object with the new loci track added</p>
    </div>
    <div class="section level2">
    <h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a></h2>

<ul><li><p><code>focus()</code>: Identify regions of interest and zoom in on them</p></li>
<li><p><code>locate()</code>: Identify regions of interest and add them as new feature track</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Let's hunt some defense systems in marine SAGs</span></span></span>
<span class="r-in"><span><span class="co"># read the genomes</span></span></span>
<span class="r-in"><span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_tracks.html">read_seqs</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"gorg/gorg.fna.fai"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'fai' with `read_fai()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: gorg.fna [/home/runner/work/_temp/Library/gggenomes/extdata/gorg/gorg.fna.fai]</span>
<span class="r-in"><span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="va">s0</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="co"># strip trailing number from contigs to get bins</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>bin_id <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">seq_id</span>, <span class="st">"_\\d+$"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># gene annotations from prokka</span></span></span>
<span class="r-in"><span><span class="va">g0</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_tracks.html">read_feats</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"gorg/gorg.gff.xz"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_gff3()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: gorg [/home/runner/work/_temp/Library/gggenomes/extdata/gorg/gorg.gff.xz]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> eC_number -&gt; e_c_number</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source       type      n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> Aragorn:1.2  tRNA     62</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> Prodigal:2.6 CDS    <span style="text-decoration: underline;">3</span>268</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> barrnap:0.6  rRNA     10</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># best hits to the PADS Arsenal database of prokaryotic defense-system genes</span></span></span>
<span class="r-in"><span><span class="co"># $ mmseqs easy-search gorg.fna pads-arsenal-v1-prf gorg-pads-defense.o6 /tmp \</span></span></span>
<span class="r-in"><span><span class="co">#     --greedy-best-hits</span></span></span>
<span class="r-in"><span><span class="va">f0</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_tracks.html">read_feats</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"gorg/gorg-pads-defense.o6"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'blast' with `read_blast()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: gorg-pads-defense [/home/runner/work/_temp/Library/gggenomes/extdata/gorg/gorg-pads-defense.o6]</span>
<span class="r-in"><span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="va">f0</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="co"># parser system/gene info</span></span></span>
<span class="r-in"><span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">seq_id2</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seq_id2"</span>, <span class="st">"system"</span>, <span class="st">"gene"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">evalue</span> <span class="op">&lt;</span> <span class="fl">1e-10</span>, <span class="co"># get rid of some spurious hits</span></span></span>
<span class="r-in"><span>    <span class="co"># and let's focus just on a few systems for this example</span></span></span>
<span class="r-in"><span>    <span class="va">system</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CRISPR-CAS"</span>, <span class="st">"DISARM"</span>, <span class="st">"GABIJA"</span>, <span class="st">"LAMASSU"</span>, <span class="st">"THOERIS"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plot the distribution of hits across full genomes</span></span></span>
<span class="r-in"><span><span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">g0</span>, <span class="va">s1</span>, <span class="va">f1</span>, wrap <span class="op">=</span> <span class="fl">2e5</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">system</span><span class="op">)</span>, data <span class="op">=</span> <span class="fu"><a href="pull_track.html">feats</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="focus-1.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># hilight the regions containing hits</span></span></span>
<span class="r-in"><span><span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">g0</span>, <span class="va">s1</span>, <span class="va">f1</span>, wrap <span class="op">=</span> <span class="fl">2e5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">locate</span><span class="op">(</span>.track_id <span class="op">=</span> <span class="va">feats</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/identity.html" class="external-link">identity</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat.html">geom_feat</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="pull_track.html">feats</a></span><span class="op">(</span><span class="va">loci</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"plum3"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">system</span><span class="op">)</span>, data <span class="op">=</span> <span class="fu"><a href="pull_track.html">feats</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding 'loci' track. Plot with `geom_feat(data=feats(loci))`</span>
<span class="r-plt img"><img src="focus-2.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># zoom in on loci</span></span></span>
<span class="r-in"><span><span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">g0</span>, <span class="va">s1</span>, <span class="va">f1</span>, wrap <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">focus</span><span class="op">(</span>.track_id <span class="op">=</span> <span class="va">feats</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat.html">geom_feat</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">system</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat_text.html">geom_feat_tag</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Showing 36 loci with the following size distribution</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> min: 2500</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> q25: 6464</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> med: 9739</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> q75: 10714</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> max: 19561</span>
<span class="r-plt img"><img src="focus-3.png" alt="" width="960" height="480"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

