<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot genomes, features and synteny maps — gggenomes • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Plot genomes, features and synteny maps — gggenomes"><meta name="description" content="gggenomes() initializes a gggenomes-flavored ggplot object.
It is used to declare the input data for gggenomes' track system.
(See for more details on the track system, gggenomes vignette or the Details/Arguments section)"><meta property="og:description" content="gggenomes() initializes a gggenomes-flavored ggplot object.
It is used to declare the input data for gggenomes' track system.
(See for more details on the track system, gggenomes vignette or the Details/Arguments section)"><meta property="og:image" content="https://thackl.github.io/gggenomes/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gggenomes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/gggenomes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a></li>
    <li><a class="dropdown-item" href="../articles/flip.html">Advanced flip examples</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thackl/gggenomes/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Plot genomes, features and synteny maps</h1>
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/master/R/gggenomes.R" class="external-link"><code>R/gggenomes.R</code></a></small>
      <div class="d-none name"><code>gggenomes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>gggenomes()</code> initializes a gggenomes-flavored ggplot object.
It is used to declare the input data for gggenomes' track system.</p>
<p>(<em>See for more details on the track system, gggenomes vignette or the Details/Arguments section</em>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gggenomes</span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seqs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feats <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  links <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .id <span class="op">=</span> <span class="st">"file_id"</span>,</span>
<span>  spacing <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  wrap <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adjacent_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  infer_bin_id <span class="op">=</span> <span class="va">seq_id</span>,</span>
<span>  infer_start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">start</span>, <span class="va">end</span><span class="op">)</span>,</span>
<span>  infer_end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">start</span>, <span class="va">end</span><span class="op">)</span>,</span>
<span>  infer_length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">start</span>, <span class="va">end</span><span class="op">)</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clean"</span>, <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>  .layout <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genes-feats">genes, feats<a class="anchor" aria-label="anchor" href="#arg-genes-feats"></a></dt>
<dd><p>A data.frame, a list of data.frames, or a character vector
with paths to files containing gene data. Each item is added as feature
track.</p>
<p>For a single data.frame the track_id will be "genes" and "feats",
respectively. For a list, track_ids are parsed from the list names, or if
names are missing from the name of the variable containing each data.frame.
Data columns:</p><ul><li><p>required: <code>seq_id,start,end</code></p></li>
<li><p>recognized: <code>strand,bin_id,feat_id,introns</code></p></li>
</ul></dd>


<dt id="arg-seqs">seqs<a class="anchor" aria-label="anchor" href="#arg-seqs"></a></dt>
<dd><p>A data.frame or a character vector with paths to files containing
sequence data. Data columns:</p><ul><li><p>required: <code>seq_id,length</code></p></li>
<li><p>recognized: <code>bin_id,start,end,strand</code></p></li>
</ul></dd>


<dt id="arg-links">links<a class="anchor" aria-label="anchor" href="#arg-links"></a></dt>
<dd><p>A data.frame or a character vector with paths to files
containing link data. Each item is added as links track. Data columns:</p><ul><li><p>required: <code>seq_id,seq_id2</code></p></li>
<li><p>recognized: <code>start,end,bin_id,start2,end2,bin_id2,strand</code></p></li>
</ul></dd>


<dt id="arg--id">.id<a class="anchor" aria-label="anchor" href="#arg--id"></a></dt>
<dd><p>The name of the column for file labels that are created when
reading directly from files. Defaults to "file_id". Set to "bin_id" if
every file represents a different bin.</p></dd>


<dt id="arg-spacing">spacing<a class="anchor" aria-label="anchor" href="#arg-spacing"></a></dt>
<dd><p>between sequences in bases (&gt;1) or relative to longest bin (&lt;1)</p></dd>


<dt id="arg-wrap">wrap<a class="anchor" aria-label="anchor" href="#arg-wrap"></a></dt>
<dd><p>wrap bins into multiple lines with at most this many nucleotides
per lin.</p></dd>


<dt id="arg-adjacent-only">adjacent_only<a class="anchor" aria-label="anchor" href="#arg-adjacent-only"></a></dt>
<dd><p>Indicates whether links should be created between adjacent sequences/chromosomes only.
By default it is set to <code>adjacent_only = TRUE</code>. If <code>FALSE</code>, links will be created between all sequences</p>
<p>(<em>not recommended for large data sets</em>)</p></dd>


<dt id="arg-infer-length-infer-start-infer-end-infer-bin-id">infer_length, infer_start, infer_end, infer_bin_id<a class="anchor" aria-label="anchor" href="#arg-infer-length-infer-start-infer-end-infer-bin-id"></a></dt>
<dd><p>used to infer pseudo
seqs if only feats or links are provided, or if no bin_id column was
provided. The expressions are evaluated in the context of the first feat
or link track.</p>
<p>By default subregions of sequences from the first to the last feat/link
are generated. Set <code>infer_start</code> to 0 to show all sequences from their
true beginning.</p></dd>


<dt id="arg-theme">theme<a class="anchor" aria-label="anchor" href="#arg-theme"></a></dt>
<dd><p>choose a gggenomes default theme, NULL to omit.</p></dd>


<dt id="arg--layout">.layout<a class="anchor" aria-label="anchor" href="#arg--layout"></a></dt>
<dd><p>a pre-computed layout from <code><a href="layout_genomes.html">layout_genomes()</a></code>. Useful for
developmental purposes.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>additional parameters, passed to layout</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>gggenomes-flavored ggplot object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>gggenomes::gggenomes()</code> resembles the functionality of <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot2::ggplot()</a></code>.
It is used to construct the initial plot object, and is often followed by "+" to add components to the plot (<em>e.g. "+ geom_gene()"</em>).</p>
<p>A big difference between the two is that gggenomes has a multi-track setup (<em><code>'seqs'</code>, <code>'feats'</code>, <code>'genes'</code> and <code>'links'</code></em>).
<code>gggenomes()</code> pre-computes a layout and adds coordinates (<code>y,x,xend</code>) to each data frame prior to the actual plot construction.
This has some implications for the usage of gggenomes:</p><ul><li><p><strong>Data frames for tracks have required variables.</strong> These predefined variables are used during import
to compute x/y coordinates (<em>see arguments</em>).</p></li>
<li><p><strong>gggenomes' geoms can often be used without explicit <code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code> mappings</strong>  This works because
we always know the names of the plot variables ahead of time: they originate from the pre-computed layout,
and we can use that information to set sensible default aesthetic mappings for most cases.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Compare the genomic organization of three viral elements</span></span></span>
<span class="r-in"><span><span class="co"># EMALEs: endogenous mavirus-like elements (example data shipped with gggenomes)</span></span></span>
<span class="r-in"><span><span class="fu">gggenomes</span><span class="op">(</span><span class="va">emale_genes</span>, <span class="va">emale_seqs</span>, <span class="va">emale_tirs</span>, <span class="va">emale_ava</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># chromosomes and labels</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat.html">geom_feat</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="op">+</span> <span class="co"># terminal inverted repeats</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">strand</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"strand"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># genes</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_link.html">geom_link</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">0.15</span><span class="op">)</span> <span class="co"># synteny-blocks</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Please use `linewidth` instead.</span>
<span class="r-plt img"><img src="gggenomes-1.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># with some more information</span></span></span>
<span class="r-in"><span><span class="fu">gggenomes</span><span class="op">(</span><span class="va">emale_genes</span>, <span class="va">emale_seqs</span>, <span class="va">emale_tirs</span>, <span class="va">emale_ava</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="add_tracks.html">add_feats</a></span><span class="op">(</span><span class="va">emale_ngaros</span>, <span class="va">emale_gc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="add_tracks.html">add_clusters</a></span><span class="op">(</span><span class="va">emale_cogs</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="flip.html">sync</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_link.html">geom_link</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">0.15</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># synteny-blocks</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># chromosomes and labels</span></span></span>
<span class="r-in"><span>  <span class="co"># thistle4, salmon4, burlywood4</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat.html">geom_feat</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span>, position <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># terminal inverted repeats</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat.html">geom_feat</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="fu"><a href="pull_track.html">feats</a></span><span class="op">(</span><span class="va">emale_ngaros</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"turquoise4"</span>, alpha <span class="op">=</span> <span class="fl">.3</span>,</span></span>
<span class="r-in"><span>    position <span class="op">=</span> <span class="st">"strand"</span>, size <span class="op">=</span> <span class="fl">16</span></span></span>
<span class="r-in"><span>  <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_feat_text.html">geom_feat_note</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">type</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="fu"><a href="pull_track.html">feats</a></span><span class="op">(</span><span class="va">emale_ngaros</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    position <span class="op">=</span> <span class="st">"strand"</span>, nudge_y <span class="op">=</span> <span class="fl">.3</span></span></span>
<span class="r-in"><span>  <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">cluster_id</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"strand"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># genes</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_wiggle.html">geom_wiggle</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>z <span class="op">=</span> <span class="va">score</span>, linetype <span class="op">=</span> <span class="st">"GC-content"</span><span class="op">)</span>, <span class="fu"><a href="pull_track.html">feats</a></span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    fill <span class="op">=</span> <span class="st">"lavenderblush4"</span>, position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_nudge.html" class="external-link">position_nudge</a></span><span class="op">(</span>y <span class="op">=</span> <span class="op">-</span><span class="fl">.2</span><span class="op">)</span>, height <span class="op">=</span> <span class="fl">.2</span></span></span>
<span class="r-in"><span>  <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span><span class="st">"Conserved genes"</span>, palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.value <span class="op">=</span> <span class="st">"cornsilk3"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(feat_id)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flipping: Cflag_017B,BVI_069,BVI_008A</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> wiggle bounds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> mid:  0.38</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> low:  0.2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> high: 0.76</span>
<span class="r-plt img"><img src="gggenomes-2.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># initialize plot directly from files</span></span></span>
<span class="r-in"><span><span class="fu">gggenomes</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.gff"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.gff"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales-tirs.gff"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.paf"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_feat.html">geom_feat</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_link.html">geom_link</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_gff3()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Note -&gt; note</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type      n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> MFG    CDS     143</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_gff3()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales-tirs [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales-tirs.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type              n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> MFG    repeat_region    12</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'paf' with `read_paf()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.paf]</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Found tags in max_tags column, you should increase max_tags to</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Read and tidied up a .paf file with 10 optional tag fields:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> NM, ms, AS, nn, tp, cm, s1, de, rl, cg</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Note: warnings about fewer than expected columns are expected for this format.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Note: .paf files use 0-based coordinate starts - transforming to 1-based</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_seq_len()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.gff]</span>
<span class="r-plt img"><img src="gggenomes-3.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># multi-contig genomes wrap to fixed width</span></span></span>
<span class="r-in"><span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_tracks.html">read_seqs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"cafeteria"</span><span class="op">)</span>, <span class="st">"Cr.*\\.fa.fai$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'fai' with `read_fai()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: CrBVI.fa [/home/runner/work/_temp/Library/gggenomes/extdata/cafeteria/CrBVI.fa.fai]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: CrCflag.fa [/home/runner/work/_temp/Library/gggenomes/extdata/cafeteria/CrCflag.fa.fai]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: CrE410P.fa [/home/runner/work/_temp/Library/gggenomes/extdata/cafeteria/CrE410P.fa.fai]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: CrRCC970.fa [/home/runner/work/_temp/Library/gggenomes/extdata/cafeteria/CrRCC970.fa.fai]</span>
<span class="r-in"><span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="va">s0</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">length</span> <span class="op">&gt;</span> <span class="fl">5e5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gggenomes</span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">s1</span>, infer_bin_id <span class="op">=</span> <span class="va">file_id</span>, wrap <span class="op">=</span> <span class="fl">5e6</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="gggenomes-4.png" alt="" width="960" height="480"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

