<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://thackl.github.io/gggenomes/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gggenomes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/gggenomes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a></li>
    <li><a class="dropdown-item" href="../articles/flip.html">Advanced flip examples</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thackl/gggenomes/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="set-up-a-plot">Set up a plot<a class="anchor" aria-label="anchor" href="#set-up-a-plot"></a></h2>

      <div class="section-desc"><p>Construct a plot combining multiple tracks</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gggenomes.html">gggenomes()</a></code>

        </dt>
        <dd>Plot genomes, features and synteny maps</dd>
      </dl></div><div class="section level2">
      <h2 id="handle-tracks">Handle tracks<a class="anchor" aria-label="anchor" href="#handle-tracks"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="track_info.html">track_info()</a></code>

        </dt>
        <dd>Basic info on tracks in a gggenomes object</dd>
      </dl><dl><dt>

          <code><a href="add_seqs.html">add_seqs()</a></code>

        </dt>
        <dd>Add seqs</dd>
      </dl><dl><dt>

          <code><a href="add_tracks.html">add_feats()</a></code> <code><a href="add_tracks.html">add_links()</a></code> <code><a href="add_tracks.html">add_subfeats()</a></code> <code><a href="add_tracks.html">add_sublinks()</a></code> <code><a href="add_tracks.html">add_clusters()</a></code>

        </dt>
        <dd>Add different types of tracks</dd>
      </dl><dl><dt>

          <code><a href="set_class.html">set_class()</a></code> <code><a href="set_class.html">add_class()</a></code> <code><a href="set_class.html">strip_class()</a></code>

        </dt>
        <dd>Modify object class attriutes</dd>
      </dl><dl><dt>

          <code><a href="pull_track.html">feats()</a></code> <code><a href="pull_track.html">feats0()</a></code> <code><a href="pull_track.html">genes()</a></code> <code><a href="pull_track.html">links()</a></code> <code><a href="pull_track.html">seqs()</a></code> <code><a href="pull_track.html">bins()</a></code> <code><a href="pull_track.html">track()</a></code> <code><a href="pull_track.html">pull_feats()</a></code> <code><a href="pull_track.html">pull_genes()</a></code> <code><a href="pull_track.html">pull_links()</a></code> <code><a href="pull_track.html">pull_seqs()</a></code> <code><a href="pull_track.html">pull_bins()</a></code> <code><a href="pull_track.html">pull_track()</a></code>

        </dt>
        <dd>Use tracks inside and outside <code>geom_*</code> calls</dd>
      </dl></div><div class="section level2">
      <h2 id="draw-geoms-positions">Draw geoms &amp; positions<a class="anchor" aria-label="anchor" href="#draw-geoms-positions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="geom_seq.html">geom_seq()</a></code>

        </dt>
        <dd>draw seqs</dd>
      </dl><dl><dt>

          <code><a href="geom_gene.html">geom_gene()</a></code>

        </dt>
        <dd>Draw gene models</dd>
      </dl><dl><dt>

          <code><a href="geom_feat.html">geom_feat()</a></code>

        </dt>
        <dd>Draw feats</dd>
      </dl><dl><dt>

          <code><a href="geom_link.html">geom_link()</a></code> <code><a href="geom_link.html">geom_link_line()</a></code>

        </dt>
        <dd>Draw links between genomes</dd>
      </dl><dl><dt>

          <code><a href="geom_wiggle.html">geom_coverage()</a></code> <code><a href="geom_wiggle.html">geom_wiggle()</a></code>

        </dt>
        <dd>Draw wiggle ribbons or lines</dd>
      </dl><dl><dt>

          <code><a href="geom_feat_text.html">geom_feat_text()</a></code> <code><a href="geom_feat_text.html">geom_feat_tag()</a></code> <code><a href="geom_feat_text.html">geom_feat_note()</a></code> <code><a href="geom_feat_text.html">geom_gene_text()</a></code> <code><a href="geom_feat_text.html">geom_gene_tag()</a></code> <code><a href="geom_feat_text.html">geom_gene_note()</a></code>

        </dt>
        <dd>Add text to genes, features, etc.</dd>
      </dl><dl><dt>

          <code><a href="geom_bin_label.html">geom_bin_label()</a></code>

        </dt>
        <dd>Draw bin labels</dd>
      </dl><dl><dt>

          <code><a href="geom_gene_label.html">geom_gene_label()</a></code> <code><a href="geom_gene_label.html">geom_feat_label()</a></code> <code><a href="geom_gene_label.html">geom_link_label()</a></code>

        </dt>
        <dd>Draw feat/link labels</dd>
      </dl><dl><dt>

          <code><a href="geom_seq_break.html">geom_seq_break()</a></code>

        </dt>
        <dd>Decorate truncated sequences</dd>
      </dl><dl><dt>

          <code><a href="geom_seq_label.html">geom_seq_label()</a></code>

        </dt>
        <dd>Draw seq labels</dd>
      </dl><dl><dt>

          <code><a href="geom_variant.html">geom_variant()</a></code>

        </dt>
        <dd>Draw place of mutation</dd>
      </dl><dl><dt>

          <code><a href="position_strand.html">position_strand()</a></code> <code><a href="position_strand.html">position_pile()</a></code> <code><a href="position_strand.html">position_strandpile()</a></code> <code><a href="position_strand.html">position_sixframe()</a></code>

        </dt>
        <dd>Stack features</dd>
      </dl><dl><dt>

          <code><a href="position_variant.html">position_variant()</a></code>

        </dt>
        <dd>Plot types of mutations with different offsets</dd>
      </dl><dl><dt>

          <code><a href="scale_x_bp.html">scale_x_bp()</a></code> <code><a href="scale_x_bp.html">label_bp()</a></code>

        </dt>
        <dd>X-scale for genomic data</dd>
      </dl><dl><dt>

          <code><a href="scale_color_variant.html">scale_color_variant()</a></code> <code><a href="scale_color_variant.html">scale_shape_variant()</a></code>

        </dt>
        <dd>Default colors and shapes for mutation types.</dd>
      </dl><dl><dt>

          <code><a href="theme_gggenomes_clean.html">theme_gggenomes_clean()</a></code>

        </dt>
        <dd>gggenomes default theme</dd>
      </dl></div><div class="section level2">
      <h2 id="manipulate-plot-data">Manipulate plot &amp; data<a class="anchor" aria-label="anchor" href="#manipulate-plot-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="pick.html">pick()</a></code> <code><a href="pick.html">pick_seqs()</a></code> <code><a href="pick.html">pick_seqs_within()</a></code> <code><a href="pick.html">pick_by_tree()</a></code>

        </dt>
        <dd>Pick bins and seqs by name or position</dd>
      </dl><dl><dt>

          <code><a href="flip.html">flip()</a></code> <code><a href="flip.html">flip_seqs()</a></code> <code><a href="flip.html">sync()</a></code>

        </dt>
        <dd>Flip bins and sequences</dd>
      </dl><dl><dt>

          <code><a href="shift.html">shift()</a></code>

        </dt>
        <dd>Shift bins left/right</dd>
      </dl><dl><dt>

          <code><a href="focus.html">focus()</a></code> <code><a href="focus.html">locate()</a></code>

        </dt>
        <dd>Show features and regions of interest</dd>
      </dl></div><div class="section level2">
      <h2 id="read-data">Read data<a class="anchor" aria-label="anchor" href="#read-data"></a></h2>

      <div class="section-desc"><p>Read common bioinformatics file types</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="read_tracks.html">read_feats()</a></code> <code><a href="read_tracks.html">read_subfeats()</a></code> <code><a href="read_tracks.html">read_links()</a></code> <code><a href="read_tracks.html">read_sublinks()</a></code> <code><a href="read_tracks.html">read_seqs()</a></code>

        </dt>
        <dd>Read files in various standard formats (FASTA, GFF3, GBK, BED, BLAST, ...) into track tables</dd>
      </dl><dl><dt>

          <code><a href="read_alitv.html">read_alitv()</a></code>

        </dt>
        <dd>Read AliTV .json file</dd>
      </dl><dl><dt>

          <code><a href="read_bed.html">read_bed()</a></code>

        </dt>
        <dd>Read a BED file</dd>
      </dl><dl><dt>

          <code><a href="read_blast.html">read_blast()</a></code>

        </dt>
        <dd>Read BLAST tab-separated output</dd>
      </dl><dl><dt>

          <code><a href="read_context.html">read_context()</a></code>

        </dt>
        <dd>Read files in different contexts</dd>
      </dl><dl><dt>

          <code><a href="read_gbk.html">read_gbk()</a></code>

        </dt>
        <dd>Read genbank files</dd>
      </dl><dl><dt>

          <code><a href="read_gff3.html">read_gff3()</a></code>

        </dt>
        <dd>Read features from GFF3 (and with some limitations GFF2/GTF) files</dd>
      </dl><dl><dt>

          <code><a href="read_paf.html">read_paf()</a></code>

        </dt>
        <dd>Read a .paf file (minimap/minimap2).</dd>
      </dl><dl><dt>

          <code><a href="read_seq_len.html">read_seq_len()</a></code> <code><a href="read_seq_len.html">read_fai()</a></code>

        </dt>
        <dd>Read sequence index</dd>
      </dl><dl><dt>

          <code><a href="read_vcf.html">read_vcf()</a></code>

        </dt>
        <dd>Read a VCF file</dd>
      </dl><dl><dt>

          <code><a href="swap_query.html">swap_query()</a></code>

        </dt>
        <dd>Swap query and subject in blast-like feature tables</dd>
      </dl><dl><dt>

          <code><a href="ex.html">ex()</a></code>

        </dt>
        <dd>Get path to gggenomes example files</dd>
      </dl><dl><dt>

          <code><a href="def_formats.html">def_formats()</a></code>

        </dt>
        <dd>Defined file formats and extensions</dd>
      </dl><dl><dt>

          <code><a href="def_names.html">def_names()</a></code> <code><a href="def_names.html">def_types()</a></code>

        </dt>
        <dd>Default column names and types for defined formats</dd>
      </dl><dl><dt>

          <code><a href="write_gff3.html">write_gff3()</a></code>

        </dt>
        <dd>Write a gff3 file from a tidy table</dd>
      </dl></div><div class="section level2">
      <h2 id="under-the-hood">Under the hood<a class="anchor" aria-label="anchor" href="#under-the-hood"></a></h2>

      <div class="section-desc"><p>Functions that power gggenomes under the hood. Useful for developers and power-users.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="layout.html">layout()</a></code>

        </dt>
        <dd>Re-layout a genome layout</dd>
      </dl><dl><dt>

          <code><a href="layout_genomes.html">layout_genomes()</a></code>

        </dt>
        <dd>Layout genomes</dd>
      </dl><dl><dt>

          <code><a href="layout_seqs.html">layout_seqs()</a></code>

        </dt>
        <dd>Layout sequences</dd>
      </dl><dl><dt>

          <code><a href="get_seqs.html">get_seqs()</a></code> <code><a href="get_seqs.html">set_seqs()</a></code>

        </dt>
        <dd>Get/set the seqs track</dd>
      </dl><dl><dt>

          <code><a href="track_ids.html">track_ids()</a></code>

        </dt>
        <dd>Named vector of track ids and types</dd>
      </dl><dl><dt>

          <code><a href="track_info.html">track_info()</a></code>

        </dt>
        <dd>Basic info on tracks in a gggenomes object</dd>
      </dl><dl><dt>

          <code><a href="require_vars.html">require_vars()</a></code>

        </dt>
        <dd>Require variables in an object</dd>
      </dl><dl><dt>

          <code><a href="vars_track.html">vars_track()</a></code>

        </dt>
        <dd>Tidyselect track variables</dd>
      </dl><dl><dt>

          <code><a href="swap_if.html">swap_if()</a></code>

        </dt>
        <dd>Swap values of two columns based on a condition</dd>
      </dl><dl><dt>

          <code><a href="split_by.html">split_by()</a></code>

        </dt>
        <dd>Split by key preserving order</dd>
      </dl><dl><dt>

          <code><a href="in_range.html">in_range()</a></code>

        </dt>
        <dd>Do numeric values fall into specified ranges?</dd>
      </dl><dl><dt>

          <code><a href="width.html">width()</a></code> <code><a href="width.html">width0()</a></code>

        </dt>
        <dd>The width of a range</dd>
      </dl><dl><dt>

          <code><a href="introduce.html">introduce()</a></code>

        </dt>
        <dd>Introduce non-existing columns</dd>
      </dl><dl><dt>

          <code><a href="qw.html">qw()</a></code> <code><a href="qw.html">qc()</a></code>

        </dt>
        <dd>Create a vector from unquoted words.</dd>
      </dl><dl><dt>

          <code><a href="GeomFeatText.html">GeomFeatText</a></code>

        </dt>
        <dd>Geom for feature text</dd>
      </dl><dl><dt>

          <code><a href="check_strand.html">check_strand()</a></code>

        </dt>
        <dd>Check strand</dd>
      </dl><dl><dt>

          <code><a href="combine_strands.html">combine_strands()</a></code>

        </dt>
        <dd>Combine strands</dd>
      </dl><dl><dt>

          <code><a href="drop_feat_layout.html">drop_feat_layout()</a></code>

        </dt>
        <dd>Drop feature layout</dd>
      </dl><dl><dt>

          <code><a href="drop_layout.html">drop_layout()</a></code>

        </dt>
        <dd>Drop a genome layout</dd>
      </dl><dl><dt>

          <code><a href="drop_link_layout.html">drop_link_layout()</a></code>

        </dt>
        <dd>Drop a link layout</dd>
      </dl><dl><dt>

          <code><a href="drop_seq_layout.html">drop_seq_layout()</a></code>

        </dt>
        <dd>Drop a seq layout</dd>
      </dl><dl><dt>

          <code><a href="flip_strand.html">flip_strand()</a></code>

        </dt>
        <dd>Flip strand</dd>
      </dl><dl><dt>

          <code><a href="if_reverse.html">if_reverse()</a></code>

        </dt>
        <dd>Vectorised if_else based on strandedness</dd>
      </dl><dl><dt>

          <code><a href="is_reverse.html">is_reverse()</a></code>

        </dt>
        <dd>Check whether strand is reverse</dd>
      </dl><dl><dt>

          <code><a href="strand_chr.html">strand_chr()</a></code>

        </dt>
        <dd>Convert strand to character</dd>
      </dl><dl><dt>

          <code><a href="strand_int.html">strand_int()</a></code>

        </dt>
        <dd>Convert strand to integer</dd>
      </dl><dl><dt>

          <code><a href="strand_lgl.html">strand_lgl()</a></code>

        </dt>
        <dd>Convert strand to logical</dd>
      </dl><dl><dt>

          <code><a href="unnest_exons.html">unnest_exons()</a></code>

        </dt>
        <dd>Unnest exons</dd>
      </dl></div><div class="section level2">
      <h2 id="data-sets">Data sets<a class="anchor" aria-label="anchor" href="#data-sets"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="emale_ava.html">emale_ava</a></code>

        </dt>
        <dd>All-versus-all whole genome alignments of 6 EMALE genomes</dd>
      </dl><dl><dt>

          <code><a href="emale_cogs.html">emale_cogs</a></code>

        </dt>
        <dd>Clusters of orthologs of 6 EMALE proteomes</dd>
      </dl><dl><dt>

          <code><a href="emale_gc.html">emale_gc</a></code>

        </dt>
        <dd>Relative GC-content along 6 EMALE genomes</dd>
      </dl><dl><dt>

          <code><a href="emale_genes.html">emale_genes</a></code>

        </dt>
        <dd>Gene annotations if 6 EMALE genomes (endogenous virophages)</dd>
      </dl><dl><dt>

          <code><a href="emale_ngaros.html">emale_ngaros</a></code>

        </dt>
        <dd>Integrated Ngaro retrotransposons of 6 EMALE genomes</dd>
      </dl><dl><dt>

          <code><a href="emale_prot_ava.html">emale_prot_ava</a></code>

        </dt>
        <dd>All-versus-all alignments 6 EMALE proteomes</dd>
      </dl><dl><dt>

          <code><a href="emale_seqs.html">emale_seqs</a></code>

        </dt>
        <dd>Sequence index of 6 EMALE genomes (endogenous virophages)</dd>
      </dl><dl><dt>

          <code><a href="emale_tirs.html">emale_tirs</a></code>

        </dt>
        <dd>Terminal inverted repeats of 6 EMALE genomes</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

