<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
    <li>
      <a href="../articles/flip.html">Advanced flip examples</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thackl/gggenomes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="set-up-a-plot">Set up a plot <a href="#set-up-a-plot" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Construct a plot combining multiple tracks</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gggenomes.html">gggenomes()</a></code> </p>
        </td>
        <td><p>Plot genomes, features and synteny maps</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="handle-tracks">Handle tracks <a href="#handle-tracks" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="track_info.html">track_info()</a></code> </p>
        </td>
        <td><p>Basic info on tracks in a gggenomes object</p></td>
      </tr><tr><td>
          <p><code><a href="add_seqs.html">add_seqs()</a></code> </p>
        </td>
        <td><p>Add seqs</p></td>
      </tr><tr><td>
          <p><code><a href="add_tracks.html">add_feats()</a></code> <code><a href="add_tracks.html">add_links()</a></code> <code><a href="add_tracks.html">add_subfeats()</a></code> <code><a href="add_tracks.html">add_sublinks()</a></code> <code><a href="add_tracks.html">add_clusters()</a></code> </p>
        </td>
        <td><p>Add different types of tracks</p></td>
      </tr><tr><td>
          <p><code><a href="set_class.html">set_class()</a></code> <code><a href="set_class.html">add_class()</a></code> <code><a href="set_class.html">strip_class()</a></code> </p>
        </td>
        <td><p>Modify object class attriutes</p></td>
      </tr><tr><td>
          <p><code><a href="pull_track.html">feats()</a></code> <code><a href="pull_track.html">feats0()</a></code> <code><a href="pull_track.html">genes()</a></code> <code><a href="pull_track.html">links()</a></code> <code><a href="pull_track.html">seqs()</a></code> <code><a href="pull_track.html">bins()</a></code> <code><a href="pull_track.html">track()</a></code> <code><a href="pull_track.html">pull_feats()</a></code> <code><a href="pull_track.html">pull_genes()</a></code> <code><a href="pull_track.html">pull_links()</a></code> <code><a href="pull_track.html">pull_seqs()</a></code> <code><a href="pull_track.html">pull_bins()</a></code> <code><a href="pull_track.html">pull_track()</a></code> </p>
        </td>
        <td><p>Use tracks inside and outside <code>geom_*</code> calls</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="draw-geoms-positions">Draw geoms &amp; positions <a href="#draw-geoms-positions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="geom_seq.html">geom_seq()</a></code> </p>
        </td>
        <td><p>draw seqs</p></td>
      </tr><tr><td>
          <p><code><a href="geom_gene.html">geom_gene()</a></code> </p>
        </td>
        <td><p>Draw gene models</p></td>
      </tr><tr><td>
          <p><code><a href="geom_feat.html">geom_feat()</a></code> </p>
        </td>
        <td><p>Draw feats</p></td>
      </tr><tr><td>
          <p><code><a href="geom_link.html">geom_link()</a></code> <code><a href="geom_link.html">geom_link_line()</a></code> </p>
        </td>
        <td><p>Draw links between genomes</p></td>
      </tr><tr><td>
          <p><code><a href="geom_wiggle.html">geom_coverage()</a></code> <code><a href="geom_wiggle.html">geom_wiggle()</a></code> </p>
        </td>
        <td><p>Draw wiggle ribbons or lines</p></td>
      </tr><tr><td>
          <p><code><a href="geom_feat_text.html">geom_feat_text()</a></code> <code><a href="geom_feat_text.html">geom_feat_tag()</a></code> <code><a href="geom_feat_text.html">geom_feat_note()</a></code> <code><a href="geom_feat_text.html">geom_gene_text()</a></code> <code><a href="geom_feat_text.html">geom_gene_tag()</a></code> <code><a href="geom_feat_text.html">geom_gene_note()</a></code> </p>
        </td>
        <td><p>Add text to genes, features, etc.</p></td>
      </tr><tr><td>
          <p><code><a href="geom_bin_label.html">geom_bin_label()</a></code> </p>
        </td>
        <td><p>Draw bin labels</p></td>
      </tr><tr><td>
          <p><code><a href="geom_gene_label.html">geom_gene_label()</a></code> <code><a href="geom_gene_label.html">geom_feat_label()</a></code> <code><a href="geom_gene_label.html">geom_link_label()</a></code> </p>
        </td>
        <td><p>Draw feat/link labels</p></td>
      </tr><tr><td>
          <p><code><a href="geom_seq_break.html">geom_seq_break()</a></code> </p>
        </td>
        <td><p>Decorate truncated sequences</p></td>
      </tr><tr><td>
          <p><code><a href="geom_seq_label.html">geom_seq_label()</a></code> </p>
        </td>
        <td><p>Draw seq labels</p></td>
      </tr><tr><td>
          <p><code><a href="geom_variant.html">geom_variant()</a></code> </p>
        </td>
        <td><p>Draw place of mutation</p></td>
      </tr><tr><td>
          <p><code><a href="position_strand.html">position_strand()</a></code> <code><a href="position_strand.html">position_pile()</a></code> <code><a href="position_strand.html">position_strandpile()</a></code> <code><a href="position_strand.html">position_sixframe()</a></code> </p>
        </td>
        <td><p>Stack features</p></td>
      </tr><tr><td>
          <p><code><a href="position_variant.html">position_variant()</a></code> </p>
        </td>
        <td><p>Plot types of mutations with different offsets</p></td>
      </tr><tr><td>
          <p><code><a href="scale_x_bp.html">scale_x_bp()</a></code> <code><a href="scale_x_bp.html">label_bp()</a></code> </p>
        </td>
        <td><p>X-scale for genomic data</p></td>
      </tr><tr><td>
          <p><code><a href="scale_color_variant.html">scale_color_variant()</a></code> <code><a href="scale_color_variant.html">scale_shape_variant()</a></code> </p>
        </td>
        <td><p>Default colors and shapes for mutation types.</p></td>
      </tr><tr><td>
          <p><code><a href="theme_gggenomes_clean.html">theme_gggenomes_clean()</a></code> </p>
        </td>
        <td><p>gggenomes default theme</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="manipulate-plot-data">Manipulate plot &amp; data <a href="#manipulate-plot-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="pick.html">pick()</a></code> <code><a href="pick.html">pick_seqs()</a></code> <code><a href="pick.html">pick_seqs_within()</a></code> <code><a href="pick.html">pick_by_tree()</a></code> </p>
        </td>
        <td><p>Pick bins and seqs by name or position</p></td>
      </tr><tr><td>
          <p><code><a href="flip.html">flip()</a></code> <code><a href="flip.html">flip_seqs()</a></code> <code><a href="flip.html">sync()</a></code> </p>
        </td>
        <td><p>Flip bins and sequences</p></td>
      </tr><tr><td>
          <p><code><a href="shift.html">shift()</a></code> </p>
        </td>
        <td><p>Shift bins left/right</p></td>
      </tr><tr><td>
          <p><code><a href="focus.html">focus()</a></code> <code><a href="focus.html">locate()</a></code> </p>
        </td>
        <td><p>Show features and regions of interest</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="read-data">Read data <a href="#read-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Read common bioinformatics file types</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="read_tracks.html">read_feats()</a></code> <code><a href="read_tracks.html">read_subfeats()</a></code> <code><a href="read_tracks.html">read_links()</a></code> <code><a href="read_tracks.html">read_sublinks()</a></code> <code><a href="read_tracks.html">read_seqs()</a></code> </p>
        </td>
        <td><p>Read files in various standard formats (FASTA, GFF3, GBK, BED, BLAST, ...) into track tables</p></td>
      </tr><tr><td>
          <p><code><a href="read_alitv.html">read_alitv()</a></code> </p>
        </td>
        <td><p>Read AliTV .json file</p></td>
      </tr><tr><td>
          <p><code><a href="read_bed.html">read_bed()</a></code> </p>
        </td>
        <td><p>Read a BED file</p></td>
      </tr><tr><td>
          <p><code><a href="read_blast.html">read_blast()</a></code> </p>
        </td>
        <td><p>Read BLAST tab-separated output</p></td>
      </tr><tr><td>
          <p><code><a href="read_context.html">read_context()</a></code> </p>
        </td>
        <td><p>Read files in different contexts</p></td>
      </tr><tr><td>
          <p><code><a href="read_gbk.html">read_gbk()</a></code> </p>
        </td>
        <td><p>Read genbank files</p></td>
      </tr><tr><td>
          <p><code><a href="read_gff3.html">read_gff3()</a></code> </p>
        </td>
        <td><p>Read features from GFF3 (and with some limitations GFF2/GTF) files</p></td>
      </tr><tr><td>
          <p><code><a href="read_paf.html">read_paf()</a></code> </p>
        </td>
        <td><p>Read a .paf file (minimap/minimap2).</p></td>
      </tr><tr><td>
          <p><code><a href="read_seq_len.html">read_seq_len()</a></code> <code><a href="read_seq_len.html">read_fai()</a></code> </p>
        </td>
        <td><p>Read sequence index</p></td>
      </tr><tr><td>
          <p><code><a href="read_vcf.html">read_vcf()</a></code> </p>
        </td>
        <td><p>Read a VCF file</p></td>
      </tr><tr><td>
          <p><code><a href="swap_query.html">swap_query()</a></code> </p>
        </td>
        <td><p>Swap query and subject in blast-like feature tables</p></td>
      </tr><tr><td>
          <p><code><a href="ex.html">ex()</a></code> </p>
        </td>
        <td><p>Get path to gggenomes example files</p></td>
      </tr><tr><td>
          <p><code><a href="def_formats.html">def_formats()</a></code> </p>
        </td>
        <td><p>Defined file formats and extensions</p></td>
      </tr><tr><td>
          <p><code><a href="def_names.html">def_names()</a></code> <code><a href="def_names.html">def_types()</a></code> </p>
        </td>
        <td><p>Default column names and types for defined formats</p></td>
      </tr><tr><td>
          <p><code><a href="write_gff3.html">write_gff3()</a></code> </p>
        </td>
        <td><p>Write a gff3 file from a tidy table</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="under-the-hood">Under the hood <a href="#under-the-hood" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions that power gggenomes under the hood. Useful for developers and power-users.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="layout.html">layout()</a></code> </p>
        </td>
        <td><p>Re-layout a genome layout</p></td>
      </tr><tr><td>
          <p><code><a href="layout_genomes.html">layout_genomes()</a></code> </p>
        </td>
        <td><p>Layout genomes</p></td>
      </tr><tr><td>
          <p><code><a href="layout_seqs.html">layout_seqs()</a></code> </p>
        </td>
        <td><p>Layout sequences</p></td>
      </tr><tr><td>
          <p><code><a href="get_seqs.html">get_seqs()</a></code> <code><a href="get_seqs.html">set_seqs()</a></code> </p>
        </td>
        <td><p>Get/set the seqs track</p></td>
      </tr><tr><td>
          <p><code><a href="track_ids.html">track_ids()</a></code> </p>
        </td>
        <td><p>Named vector of track ids and types</p></td>
      </tr><tr><td>
          <p><code><a href="track_info.html">track_info()</a></code> </p>
        </td>
        <td><p>Basic info on tracks in a gggenomes object</p></td>
      </tr><tr><td>
          <p><code><a href="require_vars.html">require_vars()</a></code> </p>
        </td>
        <td><p>Require variables in an object</p></td>
      </tr><tr><td>
          <p><code><a href="vars_track.html">vars_track()</a></code> </p>
        </td>
        <td><p>Tidyselect track variables</p></td>
      </tr><tr><td>
          <p><code><a href="swap_if.html">swap_if()</a></code> </p>
        </td>
        <td><p>Swap values of two columns based on a condition</p></td>
      </tr><tr><td>
          <p><code><a href="split_by.html">split_by()</a></code> </p>
        </td>
        <td><p>Split by key preserving order</p></td>
      </tr><tr><td>
          <p><code><a href="in_range.html">in_range()</a></code> </p>
        </td>
        <td><p>Do numeric values fall into specified ranges?</p></td>
      </tr><tr><td>
          <p><code><a href="width.html">width()</a></code> <code><a href="width.html">width0()</a></code> </p>
        </td>
        <td><p>The width of a range</p></td>
      </tr><tr><td>
          <p><code><a href="introduce.html">introduce()</a></code> </p>
        </td>
        <td><p>Introduce non-existing columns</p></td>
      </tr><tr><td>
          <p><code><a href="qw.html">qw()</a></code> <code><a href="qw.html">qc()</a></code> </p>
        </td>
        <td><p>Create a vector from unquoted words.</p></td>
      </tr><tr><td>
          <p><code><a href="GeomFeatText.html">GeomFeatText</a></code> </p>
        </td>
        <td><p>Geom for feature text</p></td>
      </tr><tr><td>
          <p><code><a href="check_strand.html">check_strand()</a></code> </p>
        </td>
        <td><p>Check strand</p></td>
      </tr><tr><td>
          <p><code><a href="combine_strands.html">combine_strands()</a></code> </p>
        </td>
        <td><p>Combine strands</p></td>
      </tr><tr><td>
          <p><code><a href="drop_feat_layout.html">drop_feat_layout()</a></code> </p>
        </td>
        <td><p>Drop feature layout</p></td>
      </tr><tr><td>
          <p><code><a href="drop_layout.html">drop_layout()</a></code> </p>
        </td>
        <td><p>Drop a genome layout</p></td>
      </tr><tr><td>
          <p><code><a href="drop_link_layout.html">drop_link_layout()</a></code> </p>
        </td>
        <td><p>Drop a link layout</p></td>
      </tr><tr><td>
          <p><code><a href="drop_seq_layout.html">drop_seq_layout()</a></code> </p>
        </td>
        <td><p>Drop a seq layout</p></td>
      </tr><tr><td>
          <p><code><a href="flip_strand.html">flip_strand()</a></code> </p>
        </td>
        <td><p>Flip strand</p></td>
      </tr><tr><td>
          <p><code><a href="if_reverse.html">if_reverse()</a></code> </p>
        </td>
        <td><p>Vectorised if_else based on strandedness</p></td>
      </tr><tr><td>
          <p><code><a href="is_reverse.html">is_reverse()</a></code> </p>
        </td>
        <td><p>Check whether strand is reverse</p></td>
      </tr><tr><td>
          <p><code><a href="strand_chr.html">strand_chr()</a></code> </p>
        </td>
        <td><p>Convert strand to character</p></td>
      </tr><tr><td>
          <p><code><a href="strand_int.html">strand_int()</a></code> </p>
        </td>
        <td><p>Convert strand to integer</p></td>
      </tr><tr><td>
          <p><code><a href="strand_lgl.html">strand_lgl()</a></code> </p>
        </td>
        <td><p>Convert strand to logical</p></td>
      </tr><tr><td>
          <p><code><a href="unnest_exons.html">unnest_exons()</a></code> </p>
        </td>
        <td><p>Unnest exons</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-sets">Data sets <a href="#data-sets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="emale_ava.html">emale_ava</a></code> </p>
        </td>
        <td><p>All-versus-all whole genome alignments of 6 EMALE genomes</p></td>
      </tr><tr><td>
          <p><code><a href="emale_cogs.html">emale_cogs</a></code> </p>
        </td>
        <td><p>Clusters of orthologs of 6 EMALE proteomes</p></td>
      </tr><tr><td>
          <p><code><a href="emale_gc.html">emale_gc</a></code> </p>
        </td>
        <td><p>Relative GC-content along 6 EMALE genomes</p></td>
      </tr><tr><td>
          <p><code><a href="emale_genes.html">emale_genes</a></code> </p>
        </td>
        <td><p>Gene annotations if 6 EMALE genomes (endogenous virophages)</p></td>
      </tr><tr><td>
          <p><code><a href="emale_ngaros.html">emale_ngaros</a></code> </p>
        </td>
        <td><p>Integrated Ngaro retrotransposons of 6 EMALE genomes</p></td>
      </tr><tr><td>
          <p><code><a href="emale_prot_ava.html">emale_prot_ava</a></code> </p>
        </td>
        <td><p>All-versus-all alignments 6 EMALE proteomes</p></td>
      </tr><tr><td>
          <p><code><a href="emale_seqs.html">emale_seqs</a></code> </p>
        </td>
        <td><p>Sequence index of 6 EMALE genomes (endogenous virophages)</p></td>
      </tr><tr><td>
          <p><code><a href="emale_tirs.html">emale_tirs</a></code> </p>
        </td>
        <td><p>Terminal inverted repeats of 6 EMALE genomes</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

