<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Pick bins and seqs by name or position — pick • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Pick bins and seqs by name or position — pick"><meta property="og:description" content="Pick which bins and seqs to show and in what order. Uses
dplyr::select()-like syntax, which means unquoted genome names, positional
arguments and selection helpers, such as
tidyselect::starts_with() are supported. Renaming is not supported."><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
    <li>
      <a href="../articles/flip.html">Advanced flip examples</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thackl/gggenomes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Pick bins and seqs by name or position</h1>
    <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/master/R/pick.R" class="external-link"><code>R/pick.R</code></a></small>
    <div class="hidden name"><code>pick.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Pick which bins and seqs to show and in what order. Uses
<code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">dplyr::select()</a></code>-like syntax, which means unquoted genome names, positional
arguments and <a href="https://tidyselect.r-lib.org/reference/language.html" class="external-link">selection helpers</a>, such as
<code><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">tidyselect::starts_with()</a></code> are supported. Renaming is not supported.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pick</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pick_seqs</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, .bins <span class="op">=</span> <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pick_seqs_within</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, .bins <span class="op">=</span> <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pick_by_tree</span><span class="op">(</span><span class="va">x</span>, <span class="va">tree</span>, infer_bin_id <span class="op">=</span> <span class="va">.data</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>gggenomes object</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>bins/seqs to pick, select-like expression.</p></dd>


<dt id="arg--bins">.bins<a class="anchor" aria-label="anchor" href="#arg--bins"></a></dt>
<dd><p>scope for positional arguments, select-like expression, enclose
multiple arguments with <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code>!</p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>a phylogenetic tree in <a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree::ggtree</a> or <code><a href="https://rdrr.io/pkg/ape/man/ape-package.html" class="external-link">ape::ape-package</a></code>-"phylo" format.</p></dd>


<dt id="arg-infer-bin-id">infer_bin_id<a class="anchor" aria-label="anchor" href="#arg-infer-bin-id"></a></dt>
<dd><p>an expression to extract bin_ids from the tree data.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>gggenomes object with selected bins and seqs.</p>
<p>gggenomes object with selected seqs.</p>
<p>gggenomes object with selected seqs.</p>
<p>gggenomes object with seqs selected by tree order.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Use the dots to select bins or sequences (depending on function suffix), and
the <code>.bins</code> argument to set the scope for positional arguments. For example,
<code>pick_seqs(1)</code> will pick the first sequence from the first bin, while
<code>pick_seqs(1, .bins=3)</code> will pick the first sequence from the third bin.</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>

<ul><li><p><code>pick()</code>: pick bins by bin_id, positional argument (start at top)
or select-helper.</p></li>
<li><p><code>pick_seqs()</code>: pick individual seqs seq_id, positional argument (start at
top left) or select-helper.</p></li>
<li><p><code>pick_seqs_within()</code>: pick individual seqs but only modify bins containing those
seqs, keep rest as is.</p></li>
<li><p><code>pick_by_tree()</code>: align bins with the leaves in a given phylogenetic tree.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  bin_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"B"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"C"</span>, <span class="st">"C"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a1"</span>, <span class="st">"b1"</span>, <span class="st">"b2"</span>, <span class="st">"b3"</span>, <span class="st">"c1"</span>, <span class="st">"c2"</span>, <span class="st">"c3"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e4</span>, <span class="fl">6e3</span>, <span class="fl">2e3</span>, <span class="fl">1e3</span>, <span class="fl">3e3</span>, <span class="fl">3e3</span>, <span class="fl">3e3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">s0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">bin_id</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expand_limits.html" class="external-link">expand_limits</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">p</span></span></span>
<span class="r-plt img"><img src="pick-1.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># remove</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick</span><span class="op">(</span><span class="op">-</span><span class="va">B</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-2.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># select and reorder, by ID and position</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick</span><span class="op">(</span><span class="va">C</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-3.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># use helper function</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick</span><span class="op">(</span><span class="fu">starts_with</span><span class="op">(</span><span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-4.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># pick just some seqs</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_seqs</span><span class="op">(</span><span class="fl">1</span>, <span class="va">c3</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-5.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># pick with .bin scope</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_seqs</span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, .bins <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-6.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># change seqs in some bins, but keep rest as is</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_seqs_within</span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, .bins <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-7.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># same w/o scope, unaffected bins remain as is</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_seqs_within</span><span class="op">(</span><span class="va">b3</span>, <span class="va">b2</span>, <span class="va">b1</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-8.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Align sequences with and plot next to a phylogenetic tree</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># arrange multiple plots</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span> <span class="co"># plot phylogenetic trees</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">ggtree v3.10.1 For help: https://yulab-smu.top/treedata-book/</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you use the ggtree package suite in published research, please cite</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> the appropriate paper(s):</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ggtree: an R package for visualization and annotation of phylogenetic</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> trees with their covariates and other associated data. Methods in</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> S Xu, Z Dai, P Guo, X Fu, S Liu, L Zhou, W Tang, T Feng, M Chen, L</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Zhan, T Wu, E Hu, Y Jiang, X Bo, G Yu. ggtreeExtra: Compact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> visualization of richly annotated phylogenetic data. Molecular Biology</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> and Evolution. 2021, 38(9):4039-4042. doi: 10.1093/molbev/msab166</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Guangchuang Yu. Using ggtree to visualize data on tree-like structures.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Current Protocols in Bioinformatics. 2020, 69:e96. doi:10.1002/cpbi.96</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘ggtree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:gggenomes’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     flip</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load and plot a phylogenetic tree</span></span></span>
<span class="r-in"><span><span class="va">emale_mcp_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales-MCP.nwk"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">emale_mcp_tree</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html" class="external-link">geom_tiplab</a></span><span class="op">(</span>align <span class="op">=</span> <span class="cn">TRUE</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.05</span><span class="op">)</span> <span class="co"># make room for labels</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">emale_seqs</span>, genes <span class="op">=</span> <span class="va">emale_genes</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plot next to each other, but with</span></span></span>
<span class="r-in"><span><span class="co"># different order in tree and genomes</span></span></span>
<span class="r-in"><span><span class="va">t</span> <span class="op">+</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="pick-9.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># reorder genomes to match tree order</span></span></span>
<span class="r-in"><span><span class="co"># with a warning caused by mismatch in y-scale expansions</span></span></span>
<span class="r-in"><span><span class="va">t</span> <span class="op">+</span> <span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_by_tree</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tree and genomes have different y-scale expansions. This can cause slight misalignments of leaves and sequences.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Consider adding `+ scale_y_continuous(expand=c(0.01,0.7,0.01,0.7))` to the tree as a fix</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> tree: 0,0.6,0,0.6</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> bins: 0.01,0.7,0.01,0.7</span>
<span class="r-plt img"><img src="pick-10.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extra genomes are dropped with a notification</span></span></span>
<span class="r-in"><span><span class="va">emale_seqs_more</span> <span class="op">&lt;-</span> <span class="va">emale_seqs</span></span></span>
<span class="r-in"><span><span class="va">emale_seqs_more</span><span class="op">[</span><span class="fl">7</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="va">emale_seqs_more</span><span class="op">[</span><span class="fl">6</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">emale_seqs_more</span><span class="op">$</span><span class="va">seq_id</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"One more genome"</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">emale_seqs_more</span>, genes <span class="op">=</span> <span class="va">emale_genes</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">t</span> <span class="op">+</span> <span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_by_tree</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Some bin_ids are missing in the tree, will drop those from genomes.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> One more genome</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tree and genomes have different y-scale expansions. This can cause slight misalignments of leaves and sequences.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Consider adding `+ scale_y_continuous(expand=c(0.01,0.7,0.01,0.7))` to the tree as a fix</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> tree: 0,0.6,0,0.6</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> bins: 0.01,0.7,0.01,0.7</span>
<span class="r-plt img"><img src="pick-11.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="co"># no shared ids will cause an error</span></span></span>
<span class="r-in"><span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>seq_id <span class="op">=</span> <span class="st">"foo"</span>, length <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">t</span> <span class="op">+</span> <span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_by_tree</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># extra leafs in tree will cause an error</span></span></span>
<span class="r-in"><span>  <span class="va">emale_seqs_fewer</span> <span class="op">&lt;-</span> <span class="fu">slice_head</span><span class="op">(</span><span class="va">emale_seqs</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">emale_seqs_fewer</span>, genes <span class="op">=</span> <span class="va">emale_genes</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">t</span> <span class="op">+</span> <span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">pick_by_tree</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Error in pick_by_tree(., t) : </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No shared bin_ids between tree and genomes. Check your IDs.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

