<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot types of mutations with different offsets — position_variant • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Plot types of mutations with different offsets — position_variant"><meta name="description" content="position_variant() allows the user to plot the different mutation types (e.g. del, ins, snps) at different offsets from the base.
This can especially be useful to highlight in which regions certain types of mutations have higher prevalence.
This position adjustment is most relevant for the analysis/visualization of VCF files with the function geom_variant()."><meta property="og:description" content="position_variant() allows the user to plot the different mutation types (e.g. del, ins, snps) at different offsets from the base.
This can especially be useful to highlight in which regions certain types of mutations have higher prevalence.
This position adjustment is most relevant for the analysis/visualization of VCF files with the function geom_variant()."><meta property="og:image" content="https://thackl.github.io/gggenomes/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gggenomes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/gggenomes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a></li>
    <li><a class="dropdown-item" href="../articles/flip.html">Advanced flip examples</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thackl/gggenomes/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Plot types of mutations with different offsets</h1>
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/main/R/position_variant.R" class="external-link"><code>R/position_variant.R</code></a></small>
      <div class="d-none name"><code>position_variant.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>position_variant()</code> allows the user to plot the different mutation types (e.g. del, ins, snps) at different offsets from the base.
This can especially be useful to highlight in which regions certain types of mutations have higher prevalence.
This position adjustment is most relevant for the analysis/visualization of VCF files with the function <code><a href="geom_variant.html">geom_variant()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">position_variant</span><span class="op">(</span>offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>del <span class="op">=</span> <span class="fl">0.1</span>, snp <span class="op">=</span> <span class="fl">0</span>, ins <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span>, base <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-offset">offset<a class="anchor" aria-label="anchor" href="#arg-offset"></a></dt>
<dd><p>Shifts the data up/down based on the type of mutation.
By default <code>offset = c(del=0.1, snp=0, ins=-0.1)</code>.
The user can supply an own vector to <code>offset</code> to indicate at which offsets the different mutation types should be plotted.
<em>Types of mutations that have not been specified within the vector, will be plotted with an offset of 0.</em></p></dd>


<dt id="arg-base">base<a class="anchor" aria-label="anchor" href="#arg-base"></a></dt>
<dd><p>How to align the offsets relative to the sequence. At base = 0, plotting of the offsets starts
from the sequence. <code>base</code> thus moves the entire feature up/down.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggproto object to be used in <code><a href="geom_variant.html">geom_variant()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Creation of example data.</span></span></span>
<span class="r-in"><span><span class="va">testposition</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ins"</span>, <span class="st">"snp"</span>, <span class="st">"snp"</span>, <span class="st">"del"</span>, <span class="st">"del"</span>, <span class="st">"snp"</span>, <span class="st">"snp"</span>, <span class="st">"ins"</span>, <span class="st">"snp"</span>, <span class="st">"ins"</span>, <span class="st">"snp"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">35</span>, <span class="fl">40</span>, <span class="fl">60</span>, <span class="fl">65</span>, <span class="fl">90</span>, <span class="fl">90</span>, <span class="fl">100</span>, <span class="fl">120</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  end <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fl">11</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">testseq</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  seq_id <span class="op">=</span> <span class="st">"A"</span>,</span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>  end <span class="op">=</span> <span class="fl">150</span>,</span></span>
<span class="r-in"><span>  length <span class="op">=</span> <span class="va">end</span> <span class="op">-</span> <span class="va">start</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">testseq</span>, feats <span class="op">=</span> <span class="va">testposition</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># This first plot shows what is being plotted when only geom_variant is called</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="position_variant-1.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Next lets use position_variant, and change the shape aesthetic by column `type`</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">type</span><span class="op">)</span>, position <span class="op">=</span> <span class="fu">position_variant</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="position_variant-2.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Now lets create a plot with different offsets by inserting a self-created vector.</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">type</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  position <span class="op">=</span> <span class="fu">position_variant</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>del <span class="op">=</span> <span class="fl">0.4</span>, ins <span class="op">=</span> <span class="op">-</span><span class="fl">0.4</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="scale_color_variant.html">scale_shape_variant</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Some mutation types are not mentioned within the offset argument. These types will have an offset of 0 by default</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No shared levels found between `names(values)` of the manual scale and the</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> data's <span style="color: #00BB00;">shape</span> values.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No shared levels found between `names(values)` of the manual scale and the</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> data's <span style="color: #00BB00;">shape</span> values.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 11 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_variant()`).</span>
<span class="r-plt img"><img src="position_variant-3.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Changing the base will shift all points up/down relatively from the sequence.</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">type</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  position <span class="op">=</span> <span class="fu">position_variant</span><span class="op">(</span>base <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="position_variant-4.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

