<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read files in various standard formats (FASTA, GFF3, GBK, BED, BLAST, ...) into track tables — read_tracks • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Read files in various standard formats (FASTA, GFF3, GBK, BED, BLAST, ...) into track tables — read_tracks"><meta property="og:description" content="Convenience functions to read sequences, features or links from various
bioinformatics file formats, such as FASTA, GFF3, Genbank, BLAST tabular
output, etc. See def_formats() for full list. File formats and the
corresponding read-functions are automatically determined based on file
extensions. All these functions can read multiple files in the same format at
once, and combine them into a single table - useful, for example, to read a
folder of gff-files with each file containing genes of a different genome."><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
    <li>
      <a href="../articles/flip.html">Advanced flip examples</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thackl/gggenomes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read files in various standard formats (FASTA, GFF3, GBK, BED, BLAST, ...) into track tables</h1>
    <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/HEAD/R/read.R" class="external-link"><code>R/read.R</code></a>, <a href="https://github.com/thackl/gggenomes/blob/HEAD/R/read_feats.R" class="external-link"><code>R/read_feats.R</code></a>, <a href="https://github.com/thackl/gggenomes/blob/HEAD/R/read_seqs.R" class="external-link"><code>R/read_seqs.R</code></a></small>
    <div class="hidden name"><code>read_tracks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Convenience functions to read sequences, features or links from various
bioinformatics file formats, such as FASTA, GFF3, Genbank, BLAST tabular
output, etc. See <code><a href="def_formats.html">def_formats()</a></code> for full list. File formats and the
corresponding read-functions are automatically determined based on file
extensions. All these functions can read multiple files in the same format at
once, and combine them into a single table - useful, for example, to read a
folder of gff-files with each file containing genes of a different genome.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_feats</span><span class="op">(</span><span class="va">files</span>, .id <span class="op">=</span> <span class="st">"file_id"</span>, format <span class="op">=</span> <span class="cn">NULL</span>, parser <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_subfeats</span><span class="op">(</span><span class="va">files</span>, .id <span class="op">=</span> <span class="st">"file_id"</span>, format <span class="op">=</span> <span class="cn">NULL</span>, parser <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_links</span><span class="op">(</span><span class="va">files</span>, .id <span class="op">=</span> <span class="st">"file_id"</span>, format <span class="op">=</span> <span class="cn">NULL</span>, parser <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_sublinks</span><span class="op">(</span><span class="va">files</span>, .id <span class="op">=</span> <span class="st">"file_id"</span>, format <span class="op">=</span> <span class="cn">NULL</span>, parser <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_seqs</span><span class="op">(</span></span>
<span>  <span class="va">files</span>,</span>
<span>  .id <span class="op">=</span> <span class="st">"file_id"</span>,</span>
<span>  format <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parser <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parse_desc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>files</dt>
<dd><p>files to reads. Should all be of same format. In many cases,
compressed files (<code>.gz</code>, <code>.bz2</code>, <code>.xz</code>, or <code>.zip</code>) are supported.
Similarly, automatic download of remote files starting with <code>http(s)://</code> or
<code>ftp(s)://</code> works in most cases.</p></dd>


<dt>.id</dt>
<dd><p>the column with the name of the file a record was read from.
Defaults to "file_id". Set to "bin_id" if every file represents a different
bin.</p></dd>


<dt>format</dt>
<dd><p>specify a format known to gggenomes, such as <code>gff3</code>, <code>gbk</code>, ...
to overwrite automatic determination based on the file extension (see
<code><a href="def_formats.html">def_formats()</a></code> for full list).</p></dd>


<dt>parser</dt>
<dd><p>specify the name of an R function to overwrite automatic
determination based on format, e.g. <code>parser="read_tsv"</code>.</p></dd>


<dt>...</dt>
<dd><p>additional arguments passed on to the format-specific read
function called down the line.</p></dd>


<dt>parse_desc</dt>
<dd><p>turn <code>key=some value</code> pairs from <code>seq_desc</code> into <code>key</code>-named
columns and remove them from <code>seq_desc</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A gggenomes-compatible sequence, feature or link tibble</p>


<p>tibble with features</p>


<p>tibble with features</p>


<p>tibble with links</p>


<p>tibble with links</p>


<p>tibble with sequence information</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>read_feats()</code>: read files as features mapping onto
sequences.</p></li>
<li><p><code>read_subfeats()</code>: read files as subfeatures mapping onto other features</p></li>
<li><p><code>read_links()</code>: read files as links connecting sequences</p></li>
<li><p><code>read_sublinks()</code>: read files as sublinks connecting features</p></li>
<li><p><code>read_seqs()</code>: read sequence ID, description and length.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># read genes/features from a gff file</span></span></span>
<span class="r-in"><span><span class="fu">read_feats</span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"eden-utr.gff"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_gff3()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: eden-utr [/home/runner/work/_temp/Library/gggenomes/extdata/eden-utr.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Parent -&gt; parent_ids</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Target -&gt; target</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 8 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type                n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> <span style="color: #BB0000;">NA</span>     CDS                 5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> <span style="color: #BB0000;">NA</span>     TF_binding_site     1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> <span style="color: #BB0000;">NA</span>     cDNA_match          1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> <span style="color: #BB0000;">NA</span>     exon                5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> <span style="color: #BB0000;">NA</span>     five_prime_UTR      1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> <span style="color: #BB0000;">NA</span>     gene                1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">7</span> <span style="color: #BB0000;">NA</span>     mRNA                5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">8</span> <span style="color: #BB0000;">NA</span>     three_prime_UTR     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 20 × 15</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    file_id  seq_id start   end strand type     feat_id introns parent_ids source</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>000  <span style="text-decoration: underline;">9</span>000 +      gene     gene00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>000  <span style="text-decoration: underline;">1</span>012 +      TF_bind… tfbs00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>050  <span style="text-decoration: underline;">9</span>000 +      mRNA     mRNA00… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>050  <span style="text-decoration: underline;">9</span>000 +      mRNA     mRNA00… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>300  <span style="text-decoration: underline;">9</span>000 +      mRNA     mRNA00… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>300  <span style="text-decoration: underline;">9</span>000 +      mRNA     mRNA00… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>300  <span style="text-decoration: underline;">1</span>500 +      exon     exon00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>050  <span style="text-decoration: underline;">1</span>500 +      exon     exon00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [2]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> eden-utr ctg123  <span style="text-decoration: underline;">3</span>000  <span style="text-decoration: underline;">3</span>902 +      exon     exon00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [2]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> eden-utr ctg123  <span style="text-decoration: underline;">5</span>000  <span style="text-decoration: underline;">5</span>500 +      exon     exon00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [3]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> eden-utr ctg123  <span style="text-decoration: underline;">7</span>000  <span style="text-decoration: underline;">9</span>000 +      exon     exon00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [3]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>201  <span style="text-decoration: underline;">7</span>600 +      CDS      cds000… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>201  <span style="text-decoration: underline;">7</span>600 +      CDS      cds000… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> eden-utr ctg123  <span style="text-decoration: underline;">3</span>301  <span style="text-decoration: underline;">7</span>600 +      CDS      cds000… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> eden-utr ctg123  <span style="text-decoration: underline;">3</span>391  <span style="text-decoration: underline;">7</span>600 +      CDS      cds000… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>050  <span style="text-decoration: underline;">9</span>000 +      mRNA     mRNA00… <span style="color: #949494;">&lt;int&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>050  <span style="text-decoration: underline;">1</span>200 +      five_pr… feat_25 <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>201  <span style="text-decoration: underline;">7</span>600 +      CDS      cds000… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">19</span> eden-utr ctg123  <span style="text-decoration: underline;">7</span>601  <span style="text-decoration: underline;">9</span>000 +      three_p… feat_30 <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">20</span> eden-utr ctg123  <span style="text-decoration: underline;">1</span>050  <span style="text-decoration: underline;">9</span>000 +      cDNA_ma… match0… <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 5 more variables: score &lt;chr&gt;, phase &lt;chr&gt;, name &lt;chr&gt;, target &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   geom_id &lt;chr&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># read all gff files from a directory</span></span></span>
<span class="r-in"><span><span class="fu">read_feats</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/"</span><span class="op">)</span>, <span class="st">"*.gff$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_gff3()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales-ngaros [/home/runner/work/_temp/Library/gggenomes/extdata/emales//emales-ngaros.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type              n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> MFG    repeat_region     3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales-tirs [/home/runner/work/_temp/Library/gggenomes/extdata/emales//emales-tirs.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type              n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> MFG    repeat_region    12</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales//emales.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Note -&gt; note</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type      n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> MFG    CDS     143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 158 × 17</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    file_id     seq_id start   end strand type  feat_id introns parent_ids source</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> emales-nga… BVI_0… <span style="text-decoration: underline;">11</span>043 <span style="text-decoration: underline;">17</span>065 <span style="color: #BB0000;">NA</span>     repe… nBVI_0… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> emales-nga… BVI_0…  <span style="text-decoration: underline;">7</span>894 <span style="text-decoration: underline;">14</span>280 <span style="color: #BB0000;">NA</span>     repe… nBVI_0… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> emales-nga… E4-10…  <span style="text-decoration: underline;">2</span>854  <span style="text-decoration: underline;">9</span>283 <span style="color: #BB0000;">NA</span>     repe… nE4-10… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> emales-tirs BVI_0…     1   473 +      repe… BVI_00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> emales-tirs BVI_0… <span style="text-decoration: underline;">26</span>348 <span style="text-decoration: underline;">26</span>820 -      repe… BVI_00… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> emales-tirs BVI_0…     1   488 +      repe… BVI_06… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> emales-tirs BVI_0… <span style="text-decoration: underline;">26</span>321 <span style="text-decoration: underline;">26</span>808 -      repe… BVI_06… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> emales-tirs Cflag…     1  <span style="text-decoration: underline;">1</span>081 +      repe… Cflag_… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> emales-tirs Cflag… <span style="text-decoration: underline;">20</span>231 <span style="text-decoration: underline;">21</span>311 -      repe… Cflag_… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> emales-tirs E4-10…     1   319 +      repe… E4-10_… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  MFG   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 148 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 7 more variables: score &lt;chr&gt;, phase &lt;chr&gt;, name &lt;chr&gt;, geom_id &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   width &lt;chr&gt;, gc_content &lt;chr&gt;, note &lt;chr&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># read remote files</span></span></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">gbk_phages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  PSSP7 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"000/858/745/GCF_000858745.1_ViralProj15134/"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"GCF_000858745.1_ViralProj15134_genomic.gff.gz"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  PSSP3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"000/904/555/GCF_000904555.1_ViralProj195517/"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"GCF_000904555.1_ViralProj195517_genomic.gff.gz"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">read_feats</span><span class="op">(</span><span class="va">gbk_phages</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_gff3()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: PSSP7 [ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/858/745/GCF_000858745.1_ViralProj15134/GCF_000858745.1_ViralProj15134_genomic.gff.gz]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Dbxref -&gt; dbxref</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> old-name -&gt; old_name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Parent -&gt; parent_ids</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Note -&gt; note</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 4 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type                 n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> RefSeq CDS                 58</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> RefSeq gene                58</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> RefSeq region               1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> RefSeq sequence_feature     1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: PSSP3 [ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/904/555/GCF_000904555.1_ViralProj195517/GCF_000904555.1_ViralProj195517_genomic.gff.gz]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmonizing attribute names</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> ID -&gt; feat_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Dbxref -&gt; dbxref</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> collection-date -&gt; collection_date</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> isolation-source -&gt; isolation_source</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> lab-host -&gt; lab_host</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> lat-lon -&gt; lat_lon</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Name -&gt; name</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Parent -&gt; parent_ids</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> Note -&gt; note</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Features read</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 4 × 3</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   source type           n</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> RefSeq CDS           53</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> RefSeq gene          53</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> RefSeq pseudogene     3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> RefSeq region         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 228 × 36</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    file_id seq_id     start   end strand type  feat_id introns parent_ids source</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> PSSP7   NC_006882…     1 <span style="text-decoration: underline;">45</span>176 +      regi… NC_006… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">1</span>079  <span style="text-decoration: underline;">1</span>816 +      gene  gene-P… <span style="color: #949494;">&lt;int&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">1</span>079  <span style="text-decoration: underline;">1</span>816 +      CDS   cds-YP… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">1</span>845  <span style="text-decoration: underline;">2</span>423 +      gene  gene-P… <span style="color: #949494;">&lt;int&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">1</span>845  <span style="text-decoration: underline;">2</span>423 +      CDS   cds-YP… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">2</span>637  <span style="text-decoration: underline;">2</span>903 +      gene  gene-P… <span style="color: #949494;">&lt;int&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">2</span>637  <span style="text-decoration: underline;">2</span>903 +      CDS   cds-YP… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">2</span>900  <span style="text-decoration: underline;">3</span>130 +      gene  gene-P… <span style="color: #949494;">&lt;int&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">2</span>900  <span style="text-decoration: underline;">3</span>130 +      CDS   cds-YP… <span style="color: #949494;">&lt;NULL&gt;</span>  <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> PSSP7   NC_006882…  <span style="text-decoration: underline;">3</span>127  <span style="text-decoration: underline;">3</span>291 +      gene  gene-P… <span style="color: #949494;">&lt;int&gt;</span>   <span style="color: #949494;">&lt;chr [1]&gt;</span>  RefSeq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 218 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 26 more variables: score &lt;chr&gt;, phase &lt;chr&gt;, name &lt;chr&gt;, dbxref &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   gbkey &lt;chr&gt;, genome &lt;chr&gt;, mol_type &lt;chr&gt;, old_name &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   gene_biotype &lt;chr&gt;, locus_tag &lt;chr&gt;, experiment &lt;chr&gt;, product &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   protein_id &lt;chr&gt;, transl_table &lt;chr&gt;, gene &lt;chr&gt;, note &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   geom_id &lt;chr&gt;, collection_date &lt;chr&gt;, isolation_source &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   lab_host &lt;chr&gt;, lat_lon &lt;chr&gt;, strain &lt;chr&gt;, description &lt;chr&gt;, …</span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># read sequences from a fasta file.</span></span></span>
<span class="r-in"><span><span class="fu">read_seqs</span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.fna"</span><span class="op">)</span>, parse_desc <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'fasta' with `read_seq_len()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.fna]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   file_id seq_id      seq_desc                                            length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> emales  BVI_008A    emale_type=EMALE01 is_typespecies=FALSE has_tir=TR…  <span style="text-decoration: underline;">26</span>820</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> emales  BVI_069     emale_type=EMALE01 is_typespecies=FALSE has_tir=TR…  <span style="text-decoration: underline;">26</span>808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> emales  Cflag_017B  emale_type=EMALE01 is_typespecies=TRUE has_tir=TRUE  <span style="text-decoration: underline;">21</span>311</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> emales  E4-10_086   emale_type=EMALE01 is_typespecies=FALSE has_tir=TR…  <span style="text-decoration: underline;">20</span>642</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> emales  E4-10_112   emale_type=EMALE01 is_typespecies=FALSE has_tir=TR…  <span style="text-decoration: underline;">26</span>856</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> emales  RCC970_016B emale_type=EMALE01 is_typespecies=FALSE has_tir=TR…  <span style="text-decoration: underline;">20</span>152</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># read sequence info from a fasta file with `parse_desc=TRUE` (default). `key=value`</span></span></span>
<span class="r-in"><span><span class="co"># pairs are removed from `seq_desc` and parsed into columns with `key` as name</span></span></span>
<span class="r-in"><span><span class="fu">read_seqs</span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.fna"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'fasta' with `read_seq_len()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.fna]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   file_id seq_id      seq_desc length emale_type is_typespecies has_tir</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> emales  BVI_008A    <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>820 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> emales  BVI_069     <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>808 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> emales  Cflag_017B  <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">21</span>311 EMALE01    TRUE           TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> emales  E4-10_086   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>642 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> emales  E4-10_112   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>856 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> emales  RCC970_016B <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>152 EMALE01    FALSE          TRUE   </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># read sequence info from samtools/seqkit style index</span></span></span>
<span class="r-in"><span><span class="fu">read_seqs</span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.fna.seqkit.fai"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'fai' with `read_fai()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales.fna.seqkit [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.fna.seqkit.fai]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   file_id           seq_id     seq_desc length emale_type is_typespecies has_tir</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> emales.fna.seqkit BVI_008A   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>820 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> emales.fna.seqkit BVI_069    <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>808 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> emales.fna.seqkit Cflag_017B <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">21</span>311 EMALE01    TRUE           TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> emales.fna.seqkit E4-10_086  <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>642 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> emales.fna.seqkit E4-10_112  <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>856 EMALE01    FALSE          TRUE   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> emales.fna.seqkit RCC970_01… <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>152 EMALE01    FALSE          TRUE   </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># read sequence info from multiple gff file</span></span></span>
<span class="r-in"><span><span class="fu">read_seqs</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.gff"</span><span class="op">)</span>, <span class="fu"><a href="ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales-tirs.gff"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading 'gff3' with `read_seq_len()`:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales.gff]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> * file_id: emales-tirs [/home/runner/work/_temp/Library/gggenomes/extdata/emales/emales-tirs.gff]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 12 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    file_id     seq_id      seq_desc length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> emales      BVI_008A    <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>820</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> emales      BVI_069     <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> emales      Cflag_017B  <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">21</span>311</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> emales      E4-10_086   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>642</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> emales      E4-10_112   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>856</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> emales      RCC970_016B <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>152</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> emales-tirs BVI_008A    <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>820</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> emales-tirs BVI_069     <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> emales-tirs Cflag_017B  <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">21</span>311</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> emales-tirs E4-10_086   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>642</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> emales-tirs E4-10_112   <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">26</span>856</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> emales-tirs RCC970_016B <span style="color: #BB0000;">NA</span>        <span style="text-decoration: underline;">20</span>152</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

