<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read features from GFF3 (and with some limitations GFF2/GTF) files — read_gff3 • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Read features from GFF3 (and with some limitations GFF2/GTF) files — read_gff3"><meta name="description" content="Files with ##FASTA section work but result in parsing problems for all
lines of the fasta section. Just ignore those warnings, or strip the fasta
section ahead of time from the file."><meta property="og:description" content="Files with ##FASTA section work but result in parsing problems for all
lines of the fasta section. Just ignore those warnings, or strip the fasta
section ahead of time from the file."><meta property="og:image" content="https://thackl.github.io/gggenomes/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gggenomes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/gggenomes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a></li>
    <li><a class="dropdown-item" href="../articles/flip.html">Advanced flip examples</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thackl/gggenomes/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Read features from GFF3 (and with some limitations GFF2/GTF) files</h1>
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/main/R/read_gff3.R" class="external-link"><code>R/read_gff3.R</code></a></small>
      <div class="d-none name"><code>read_gff3.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Files with <code>##FASTA</code> section work but result in parsing problems for all
lines of the fasta section. Just ignore those warnings, or strip the fasta
section ahead of time from the file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_gff3</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  sources <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  infer_cds_parents <span class="op">=</span> <span class="va">is_gff2</span>,</span>
<span>  sort_exons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  col_names <span class="op">=</span> <span class="fu"><a href="def_names.html">def_names</a></span><span class="op">(</span><span class="st">"gff3"</span><span class="op">)</span>,</span>
<span>  col_types <span class="op">=</span> <span class="fu"><a href="def_names.html">def_types</a></span><span class="op">(</span><span class="st">"gff3"</span><span class="op">)</span>,</span>
<span>  keep_attr <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fix_augustus_cds <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  is_gff2 <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>Either a path to a file, a connection, or literal data
(either a single string or a raw vector).</p>
<p>Files ending in <code>.gz</code>, <code>.bz2</code>, <code>.xz</code>, or <code>.zip</code> will
be automatically uncompressed. Files starting with <code>http://</code>,
<code>https://</code>, <code>ftp://</code>, or <code>ftps://</code> will be automatically
downloaded. Remote gz files can also be automatically downloaded and
decompressed.</p>
<p>Literal data is most useful for examples and tests. To be recognised as
literal data, the input must be either wrapped with <code><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I()</a></code>, be a string
containing at least one new line, or be a vector containing at least one
string with a new line.</p>
<p>Using a value of <code><a href="https://readr.tidyverse.org/reference/clipboard.html" class="external-link">clipboard()</a></code> will read from the system clipboard.</p></dd>


<dt id="arg-sources">sources<a class="anchor" aria-label="anchor" href="#arg-sources"></a></dt>
<dd><p>only return features from these sources</p></dd>


<dt id="arg-types">types<a class="anchor" aria-label="anchor" href="#arg-types"></a></dt>
<dd><p>only return features of these types, e.g. gene, CDS, ...</p></dd>


<dt id="arg-infer-cds-parents">infer_cds_parents<a class="anchor" aria-label="anchor" href="#arg-infer-cds-parents"></a></dt>
<dd><p>infer the mRNA parent for CDS features based on
overlapping coordinates. Default TRUE for gff2/gtf, FALSE for gff3. In most
GFFs this is properly set, but sometimes this information is missing.
Generally, this is not a problem, however, geom_gene calls parse the parent
information to determine which CDS and mRNAs are part of the same gene
model. Without the parent info, mRNA and CDS are plotted as individual
features.</p></dd>


<dt id="arg-sort-exons">sort_exons<a class="anchor" aria-label="anchor" href="#arg-sort-exons"></a></dt>
<dd><p>make sure that exons/introns appear sorted. Default TRUE.
Set to FALSE to read CDS/exon order exactly as present in the file, which
is less robust, but faster and allows non-canonical splicing
(exon1-exon3-exon2).</p></dd>


<dt id="arg-col-names">col_names<a class="anchor" aria-label="anchor" href="#arg-col-names"></a></dt>
<dd><p>column names to use. Defaults to <code>def_names("gff3")</code> (see <code><a href="def_names.html">def_names</a></code>).</p></dd>


<dt id="arg-col-types">col_types<a class="anchor" aria-label="anchor" href="#arg-col-types"></a></dt>
<dd><p>column types to use. Defaults to <code>def_types("gff3")</code> (see <code><a href="def_names.html">def_types</a></code>).</p></dd>


<dt id="arg-keep-attr">keep_attr<a class="anchor" aria-label="anchor" href="#arg-keep-attr"></a></dt>
<dd><p>keep the original attributes column also after parsing
tag=value pairs into tidy columns.</p></dd>


<dt id="arg-fix-augustus-cds">fix_augustus_cds<a class="anchor" aria-label="anchor" href="#arg-fix-augustus-cds"></a></dt>
<dd><p>If true, assume Augustus gff with bad CDS IDs that
need fixing</p></dd>


<dt id="arg-is-gff-">is_gff2<a class="anchor" aria-label="anchor" href="#arg-is-gff-"></a></dt>
<dd><p>set if file is in gff2 format</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>tibble</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

