<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read features from GFF3 (and with some limitations GFF2/GTF) files — read_gff3 • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Read features from GFF3 (and with some limitations GFF2/GTF) files — read_gff3"><meta property="og:description" content="Files with ##FASTA section work but result in parsing problems for all
lines of the fasta section. Just ignore those warnings, or strip the fasta
section ahead of time from the file."><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
    <li>
      <a href="../articles/flip.html">Advanced flip examples</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thackl/gggenomes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read features from GFF3 (and with some limitations GFF2/GTF) files</h1>
    <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/master/R/read_gff3.R" class="external-link"><code>R/read_gff3.R</code></a></small>
    <div class="hidden name"><code>read_gff3.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Files with <code>##FASTA</code> section work but result in parsing problems for all
lines of the fasta section. Just ignore those warnings, or strip the fasta
section ahead of time from the file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_gff3</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  sources <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  infer_cds_parents <span class="op">=</span> <span class="va">is_gff2</span>,</span>
<span>  sort_exons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  col_names <span class="op">=</span> <span class="fu"><a href="def_names.html">def_names</a></span><span class="op">(</span><span class="st">"gff3"</span><span class="op">)</span>,</span>
<span>  col_types <span class="op">=</span> <span class="fu"><a href="def_names.html">def_types</a></span><span class="op">(</span><span class="st">"gff3"</span><span class="op">)</span>,</span>
<span>  keep_attr <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fix_augustus_cds <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  is_gff2 <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>Either a path to a file, a connection, or literal data
(either a single string or a raw vector).</p>
<p>Files ending in <code>.gz</code>, <code>.bz2</code>, <code>.xz</code>, or <code>.zip</code> will
be automatically uncompressed. Files starting with <code>http://</code>,
<code>https://</code>, <code>ftp://</code>, or <code>ftps://</code> will be automatically
downloaded. Remote gz files can also be automatically downloaded and
decompressed.</p>
<p>Literal data is most useful for examples and tests. To be recognised as
literal data, the input must be either wrapped with <code><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I()</a></code>, be a string
containing at least one new line, or be a vector containing at least one
string with a new line.</p>
<p>Using a value of <code><a href="https://readr.tidyverse.org/reference/clipboard.html" class="external-link">clipboard()</a></code> will read from the system clipboard.</p></dd>


<dt id="arg-sources">sources<a class="anchor" aria-label="anchor" href="#arg-sources"></a></dt>
<dd><p>only return features from these sources</p></dd>


<dt id="arg-types">types<a class="anchor" aria-label="anchor" href="#arg-types"></a></dt>
<dd><p>only return features of these types, e.g. gene, CDS, ...</p></dd>


<dt id="arg-infer-cds-parents">infer_cds_parents<a class="anchor" aria-label="anchor" href="#arg-infer-cds-parents"></a></dt>
<dd><p>infer the mRNA parent for CDS features based on
overlapping coordinates. Default TRUE for gff2/gtf, FALSE for gff3. In most
GFFs this is properly set, but sometimes this information is missing.
Generally, this is not a problem, however, geom_gene calls parse the parent
information to determine which CDS and mRNAs are part of the same gene
model. Without the parent info, mRNA and CDS are plotted as individual
features.</p></dd>


<dt id="arg-sort-exons">sort_exons<a class="anchor" aria-label="anchor" href="#arg-sort-exons"></a></dt>
<dd><p>make sure that exons/introns appear sorted. Default TRUE.
Set to FALSE to read CDS/exon order exactly as present in the file, which
is less robust, but faster and allows non-canonical splicing
(exon1-exon3-exon2).</p></dd>


<dt id="arg-col-names">col_names<a class="anchor" aria-label="anchor" href="#arg-col-names"></a></dt>
<dd><p>column names to use. Defaults to <code>def_names("gff3")</code> (see <code><a href="def_names.html">def_names</a></code>).</p></dd>


<dt id="arg-col-types">col_types<a class="anchor" aria-label="anchor" href="#arg-col-types"></a></dt>
<dd><p>column types to use. Defaults to <code>def_types("gff3")</code> (see <code><a href="def_names.html">def_types</a></code>).</p></dd>


<dt id="arg-keep-attr">keep_attr<a class="anchor" aria-label="anchor" href="#arg-keep-attr"></a></dt>
<dd><p>keep the original attributes column also after parsing
tag=value pairs into tidy columns.</p></dd>


<dt id="arg-fix-augustus-cds">fix_augustus_cds<a class="anchor" aria-label="anchor" href="#arg-fix-augustus-cds"></a></dt>
<dd><p>If true, assume Augustus gff with bad CDS IDs that
need fixing</p></dd>


<dt id="arg-is-gff-">is_gff2<a class="anchor" aria-label="anchor" href="#arg-is-gff-"></a></dt>
<dd><p>set if file is in gff2 format</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>tibble</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

