<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Default colors and shapes for mutation types. — scale_color_variant • gggenomes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Default colors and shapes for mutation types. — scale_color_variant"><meta property="og:description" content="The user can call upon an convenient function called scale_color_variant,
which changes the color of (SNP) points, based on their nucleotides (A, C, G, T).
By default the function uses a colorblind friendly palette, but users can manually overwrite these colors.
(Within the plotting function (e.g. geom_variant), coloring of the column should still be mentioned (aes(color = ...)).
The function scale_shape_variant changes the shape of plotted points based on the type of mutation.
The user can also manually decide which shape, each specific type of mutation should have.
By default, SNPs are diamond shaped, Deletions triangle downwards and Insertions triangle upwards.
(These default settings make most sense when using geom_variant(offset = -0.2)).
(User should still manually call which column is used for the shape aesthetic)"><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.8.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thackl/gggenomes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Default colors and shapes for mutation types.</h1>
    <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/HEAD/R/variant_scales.R" class="external-link"><code>R/variant_scales.R</code></a></small>
    <div class="hidden name"><code>scale_color_variant.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The user can call upon an convenient function called <code>scale_color_variant</code>,
which changes the color of (SNP) points, based on their nucleotides (A, C, G, T).
By default the function uses a colorblind friendly palette, but users can manually overwrite these colors.
<em>(Within the plotting function (e.g. <code>geom_variant</code>), coloring of the column should still be mentioned (<code>aes(color = ...)</code>).</em></p>
<p>The function <code>scale_shape_variant</code> changes the shape of plotted points based on the type of mutation.
The user can also manually decide which shape, each specific type of mutation should have.
By default, SNPs are diamond shaped, Deletions triangle downwards and Insertions triangle upwards.
(These default settings make most sense when using <code>geom_variant(offset = -0.2)</code>).
<em>(User should still manually call which column is used for the shape aesthetic)</em></p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scale_color_variant</span><span class="op">(</span></span>
<span>  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="st">"#e66101"</span>, C <span class="op">=</span> <span class="st">"#b2abd2"</span>, G <span class="op">=</span> <span class="st">"#5e3c99"</span>, T <span class="op">=</span> <span class="st">"#fdb863"</span><span class="op">)</span>,</span>
<span>  na.value <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scale_shape_variant</span><span class="op">(</span></span>
<span>  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNP <span class="op">=</span> <span class="fl">23</span>, Deletion <span class="op">=</span> <span class="fl">25</span>, Insertion <span class="op">=</span> <span class="fl">24</span><span class="op">)</span>,</span>
<span>  na.value <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  characters <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>values</dt>
<dd><p>A vector indicating how to color/shape different variables.
The functions <code>scale_color_variant()</code> and <code>scale_shape_variant()</code> have a default setting, which can be overwritten.</p></dd>


<dt>na.value</dt>
<dd><p>The aesthetic value (color/shape/etc.) to use for non matching values.</p></dd>


<dt>characters</dt>
<dd><p>When <code>TRUE</code>, it changes the default shapes of <code>scale_shape_variant()</code> to become the letters of the nucleotides.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Creation of example data. </span></span></span>
<span class="r-in"><span><span class="va">testposition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Insertion"</span>, <span class="st">"SNP"</span>, <span class="st">"SNP"</span>, <span class="st">"Deletion"</span>, <span class="st">"Deletion"</span>, <span class="st">"SNP"</span>, <span class="st">"SNP"</span>, <span class="st">"Insertion"</span>, <span class="st">"SNP"</span>, <span class="st">"Insertion"</span>, <span class="st">"SNP"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">35</span>, <span class="fl">40</span>, <span class="fl">60</span>, <span class="fl">65</span>, <span class="fl">90</span>, <span class="fl">90</span>, <span class="fl">100</span>, <span class="fl">120</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>ALT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AT"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>, <span class="st">"."</span>, <span class="st">"."</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"CAT"</span>, <span class="st">"G"</span>, <span class="st">"TC"</span>, <span class="st">"A"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>REF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"G"</span>, <span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>end <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fl">11</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">testseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>seq_id <span class="op">=</span> <span class="st">"A"</span>,</span></span>
<span class="r-in"><span>start <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>end <span class="op">=</span> <span class="fl">150</span>,</span></span>
<span class="r-in"><span>length <span class="op">=</span> <span class="va">end</span><span class="op">-</span><span class="va">start</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs<span class="op">=</span><span class="va">testseq</span>, feats<span class="op">=</span><span class="va">testposition</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Scale_color_variant()</span></span></span>
<span class="r-in"><span><span class="co"># Changing the color aesthetics in geom_variant: colors all mutations</span></span></span>
<span class="r-in"><span><span class="co"># (In this example, All ALT (alternative) nucleotides are being colored)</span></span></span>
<span class="r-in"><span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">ALT</span><span class="op">)</span><span class="op">)</span> </span></span>
<span class="r-plt img"><img src="scale_color_variant-1.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Color all SNPs with default colors using scale_color_variant().</span></span></span>
<span class="r-in"><span><span class="co">#(SNPs are 1 nucleotide long, other mutations such as Insertions and Deletions have either more ore less nucleotides within the ALT column and are thus not plotted)</span></span></span>
<span class="r-in"><span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">ALT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_color_variant</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scale_color_variant-2.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Manually changing colors with scale_color_variant()</span></span></span>
<span class="r-in"><span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">ALT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_color_variant</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A<span class="op">=</span><span class="st">"purple"</span>, T<span class="op">=</span><span class="st">"darkred"</span>, TC<span class="op">=</span><span class="st">"black"</span>, AT<span class="op">=</span><span class="st">"pink"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scale_color_variant-3.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Scale_shape_variant()</span></span></span>
<span class="r-in"><span><span class="co"># Changing the `shape` aesthetics in geom_variant</span></span></span>
<span class="r-in"><span><span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">type</span><span class="op">)</span>, offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scale_color_variant-4.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calling upon scale_shape_variant() to change shapes</span></span></span>
<span class="r-in"><span><span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">type</span><span class="op">)</span>, offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_shape_variant</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scale_color_variant-5.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Manually changing shapes with scale_shape_variant()</span></span></span>
<span class="r-in"><span><span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">type</span><span class="op">)</span>, offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_shape_variant</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNP<span class="op">=</span><span class="fl">14</span>, Deletion<span class="op">=</span><span class="fl">18</span>, Insertion<span class="op">=</span><span class="fl">21</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scale_color_variant-6.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Plotting (nucleotides) characters instead of shapes</span></span></span>
<span class="r-in"><span><span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">ALT</span><span class="op">)</span>, offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span>, size<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_shape_variant</span><span class="op">(</span>characters<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 5 rows containing missing values (`geom_variant()`).</span>
<span class="r-plt img"><img src="scale_color_variant-7.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Alternative way to plot nucleotides (of ALT) by using `geom=text` within `geom_variant()`</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs<span class="op">=</span><span class="va">s1</span>, feats<span class="op">=</span><span class="va">f1</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">type</span><span class="op">)</span>, offset<span class="op">=</span><span class="op">-</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span> <span class="fu">scale_shape_variant</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">ALT</span><span class="op">)</span>, geom<span class="op">=</span><span class="st">"text"</span>, offset<span class="op">=</span><span class="op">-</span><span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span>  </span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in as_tracks(feats, feats_exprs, c("seqs", names2(genes)), context = "feats"):</span> object 'f1' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Combining scale_color_variant() and scale_shape_variant()</span></span></span>
<span class="r-in"><span><span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">ALT</span>, color <span class="op">=</span> <span class="va">ALT</span><span class="op">)</span>, offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span>, size <span class="op">=</span> <span class="fl">3</span>, show.legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu"><a href="geom_variant.html">geom_variant</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">ALT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_color_variant</span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span><span class="fu">scale_shape_variant</span><span class="op">(</span>characters<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 5 rows containing missing values (`geom_variant()`).</span>
<span class="r-plt img"><img src="scale_color_variant-8.png" alt="" width="960" height="480"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

