<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot genomes, features and synteny maps — gggenomes • gggenomes</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot genomes, features and synteny maps — gggenomes" />
<meta property="og:description" content="Sequence data: `read_fai`" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot genomes, features and synteny maps</h1>
    
    <div class="hidden name"><code>gggenomes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Sequence data: `read_fai`</p>
    </div>

    <pre class="usage"><span class='fu'>gggenomes</span>(<span class='kw'>seqs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>features</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>links</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>, <span class='kw'>infer_bin_id</span> <span class='kw'>=</span> <span class='no'>seq_id</span>, <span class='kw'>infer_start</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span>(<span class='no'>start</span>, <span class='no'>end</span>),
  <span class='kw'>infer_end</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span>(<span class='no'>start</span>, <span class='no'>end</span>), <span class='kw'>infer_length</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span>(<span class='no'>start</span>, <span class='no'>end</span>),
  <span class='kw'>theme</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"clean"</span>, <span class='kw'>NULL</span>), <span class='kw'>.layout</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='fu'>layout_genomes</span>(<span class='kw'>seqs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>features</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>links</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>infer_bin_id</span> <span class='kw'>=</span> <span class='no'>seq_id</span>, <span class='kw'>infer_start</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span>(<span class='no'>start</span>, <span class='no'>end</span>),
  <span class='kw'>infer_end</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span>(<span class='no'>start</span>, <span class='no'>end</span>), <span class='kw'>infer_length</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span>(<span class='no'>start</span>, <span class='no'>end</span>), <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>seqs</th>
      <td><p>a table with sequence data (seq_id, bin_id, length)</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>a table or a list of table with gene data to be added as feature
tracks. Required columns: seq_id, bin_id, start, end.</p>
<p>For a single table, adds the track_id will be "genes". For a list, track_ids
are parsed from the list names, or if names are missing from the name of the
variable containing each table.</p></td>
    </tr>
    <tr>
      <th>features</th>
      <td><p>same as genes, but the single table track_id will default to
"features".</p></td>
    </tr>
    <tr>
      <th>links</th>
      <td><p>a table or a list of tables with link data to be added as link
tracks (columns: from, to, from_start, from_end, to_start, to_end). Same
naming scheme as for features.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>layout parameters passed on to [layout_seqs()]</p></td>
    </tr>
    <tr>
      <th>infer_length, infer_start, infer_end, infer_bin_id</th>
      <td><p>used to infer pseudo
seqs if only features or links are provided, or if no bin_id column was
provided. The expressions are evaluated in the context of the first feature
or link track.</p>
<p>By default subregions of sequences from the first to the last feature/link
are generated. Set `infer_start` to 0 to show all sequences from their
true beginning.</p></td>
    </tr>
    <tr>
      <th>theme</th>
      <td><p>choose a gggenomes default theme, NULL to omit.</p></td>
    </tr>
    <tr>
      <th>.layout</th>
      <td><p>a pre-computed layout from [layout_genomes()]. Useful for
developmental purposes.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>gggenomes-flavored ggplot object</p>
<p>gggenomes_layout object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Feature data: `read_gff`</p>
<p>Link data: `read_paf`</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Compare the genomic organization of three viral elements</span>
<span class='co'># EMALEs: endogenous mavirus-like elements (example data shipped with gggenomes)</span>
<span class='fu'>gggenomes</span>(<span class='no'>emale_seqs</span>[<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>1</span>:<span class='fl'>2</span>,<span class='fl'>6</span>),], <span class='no'>emale_genes</span>, <span class='no'>emale_tirs</span>, <span class='no'>emale_links</span>) +
  <span class='fu'><a href='geom_seq.html'>geom_seq</a></span>() + <span class='fu'><a href='geom_bin_label.html'>geom_bin_label</a></span>() +                  <span class='co'># chromosomes and labels</span>
  <span class='fu'><a href='geom_feature.html'>geom_feature</a></span>(<span class='kw'>size</span><span class='kw'>=</span><span class='fl'>8</span>) +                           <span class='co'># terminal inverted repeats</span>
  <span class='fu'><a href='geom_gene.html'>geom_gene</a></span>(<span class='fu'>aes</span>(<span class='kw'>fill</span><span class='kw'>=</span><span class='no'>strand</span>), <span class='kw'>position</span><span class='kw'>=</span><span class='st'>"strand"</span>) + <span class='co'># genes</span>
  <span class='fu'><a href='geom_link.html'>geom_link</a></span>(<span class='kw'>offset</span> <span class='kw'>=</span> <span class='fl'>0.15</span>)                         <span class='co'># synteny-blocks</span></div><div class='img'><img src='gggenomes-1.png' alt='' width='1152' height='768' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thomas Hackl.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


