<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>From a few sequences to a complex map in minutes (old version) • gggenomes</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="From a few sequences to a complex map in minutes (old version)">
<meta property="og:description" content="gggenomes">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.4.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thackl/gggenomes/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="emales_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>From a few sequences to a complex map in minutes (old version)</h1>
                        <h4 class="author">Thomas Hackl</h4>
            
            <h4 class="date">2021-09-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/master/vignettes/emales.Rmd"><code>vignettes/emales.Rmd</code></a></small>
      <div class="hidden name"><code>emales.Rmd</code></div>

    </div>

    
    
<p><strong>DISCLAIMER: I’ve created this demo with an early version of gggenomes. Not all of the code will function with current releases. However, the general workflow both with respect to the external tools and gggenomes code is still valid and this demo should therefore be understood as an hopefully inspiring guide - not as an exactly reproducible code example (Those kind of examples you can find in the examples sections of the documentation).</strong></p>
<p>This is a real-life example demonstrating the use of <code>gggenomes</code> to explore viral genomes. We start with just a bunch of viral contigs and ask: Is there anything interesting going on here?</p>
<p>We use a few bioinformatics commandline tools to run some analyzes, and visualize the results using <code>gggenomes</code>. By successively adding new data as new tracks we build a rich plot that ultimatly reveals novel insights into an exciting system.</p>
<p>This example is bundled <code><a href="https://rdrr.io/r/utils/data.html">data(package="gggenomes")</a></code> To rerun the this example including the bioinformatics analyzes download the <a href="https://github.com/thackl/gggenomes/tree/master/data-raw/emales">raw data</a> from github.</p>
<p><img src="emales/emales-p7.png"></p>
<div id="read-in-the-genomes" class="section level3">
<h3 class="hasAnchor">
<a href="#read-in-the-genomes" class="anchor"></a>Read in the genomes</h3>
<p>We start with a fasta file of 33 viral genomes. We read sequence length and some metadata from the header lines using `read<sub>fai</sub>`…</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thackl/gggenomes">gggenomes</a></span><span class="op">)</span>

<span class="co"># parse sequence length and some metadata from fasta file</span>
<span class="va">emale_seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_seq_len.html">read_fai</a></span><span class="op">(</span><span class="st">"emales.fna"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">extract</span><span class="op">(</span><span class="va">seq_desc</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"emale_type"</span>, <span class="st">"is_typespecies"</span><span class="op">)</span>, <span class="st">"=(\\S+) \\S+=(\\S+)"</span>,
    remove<span class="op">=</span><span class="cn">F</span>, convert<span class="op">=</span><span class="cn">T</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">arrange</span><span class="op">(</span><span class="va">emale_type</span>, <span class="va">length</span><span class="op">)</span>

<span class="co"># plot the genomes - first six only to keep it simple for this example</span>
<span class="va">emale_seqs_6</span> <span class="op">&lt;-</span> <span class="va">emale_seqs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="op">]</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">emale_seqs_6</span><span class="op">)</span> <span class="op">+</span>
 <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="emales/emales-p1.png"></p>
</div>
<div id="annotate-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#annotate-genes" class="anchor"></a>Annotate genes</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># https://github.com/thackl/seq-scripts</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="ex">seq-join</span> -n emales-concat <span class="op">&lt;</span> emales.fna <span class="op">&gt;</span> emales-concat.fna                       </span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co"># Annotate genes | https://github.com/hyattpd/Prodigal</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="ex">prodigal</span> -n -t emales-prodigal.train -i emales-concat.fna                                 </span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="ex">prodigal</span> -t emales-prodigal.train -i emales.fna -o emales-prodigal.gff -f gff</span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co"># A little help to clean up the prodigal gff | https://github.com/thackl/seq-scripts</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="ex">gff-clean</span> emales-prodigal.gff <span class="op">&gt;</span> emales.gff</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emale_genes</span> <span class="op">&lt;-</span> <span class="fu">read_gff</span><span class="op">(</span><span class="st">"emales.gff"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">rename</span><span class="op">(</span>feature_id<span class="op">=</span><span class="va">ID</span><span class="op">)</span> <span class="op">%&gt;%</span>                       <span class="co"># we'll need this later</span>
  <span class="fu">mutate</span><span class="op">(</span>gc_cont<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">gc_cont</span><span class="op">)</span><span class="op">)</span>             <span class="co"># per gene GC-content</span>

<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">emale_seqs_6</span>, <span class="va">emale_genes</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">gc_cont</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_distiller</span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Spectral"</span><span class="op">)</span>
<span class="va">p2</span></code></pre></div>
<p><img src="emales/emales-p2.png"></p>
</div>
<div id="find-terminal-inverted-repeats" class="section level3">
<h3 class="hasAnchor">
<a href="#find-terminal-inverted-repeats" class="anchor"></a>Find terminal inverted repeats</h3>
<p>It is known that these type of viruses often have linear genomes with terminal inverted repeats (TIRs). So let’s look for those next.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># split into one genome per file | https://bioinf.shenwei.me/seqkit/</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="ex">seqkit</span> split -i emales.fna  </span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># self-align opposite strands                        </span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="kw">for</span> <span class="ex">fna</span> in <span class="kw">`</span><span class="fu">ls</span> emales.fna.split/*.fna<span class="kw">`;</span> <span class="kw">do</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>  <span class="ex">minimap2</span> -c -B5 -O6 -E3 --rev-only <span class="va">$fna</span> <span class="va">$fna</span> <span class="op">&gt;</span> <span class="va">$fna</span>.paf<span class="kw">;</span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="kw">done</span>;</span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="fu">cat</span> emales.fna.split/*.paf <span class="op">&gt;</span> emales-tirs.paf</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># prefilter hits by minimum length and maximum divergence</span>
<span class="va">emale_tirs_paf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_paf.html">read_paf</a></span><span class="op">(</span><span class="st">"emales-tirs.paf"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">seq_id1</span> <span class="op">==</span> <span class="va">seq_id2</span> <span class="op">&amp;</span> <span class="va">start1</span> <span class="op">&lt;</span> <span class="va">start2</span> <span class="op">&amp;</span> <span class="va">map_length</span> <span class="op">&gt;</span> <span class="fl">99</span> <span class="op">&amp;</span> <span class="va">de</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span>
<span class="va">emale_tirs</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">emale_tirs_paf</span>, seq_id<span class="op">=</span><span class="va">seq_id1</span>, start<span class="op">=</span><span class="va">start1</span>, end<span class="op">=</span><span class="va">end1</span>, <span class="va">de</span><span class="op">)</span>,
  <span class="fu">select</span><span class="op">(</span><span class="va">emale_tirs_paf</span>, seq_id<span class="op">=</span><span class="va">seq_id2</span>, start<span class="op">=</span><span class="va">start2</span>, end<span class="op">=</span><span class="va">end2</span>, <span class="va">de</span><span class="op">)</span><span class="op">)</span>

<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">emale_seqs_6</span>, <span class="va">emale_genes</span>, <span class="va">emale_tirs</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_feature</span><span class="op">(</span>size<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">gc_cont</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_distiller</span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Spectral"</span><span class="op">)</span>
<span class="va">p3</span></code></pre></div>
<p><img src="emales/emales-p3.png"></p>
</div>
<div id="compare-genome-synteny" class="section level3">
<h3 class="hasAnchor">
<a href="#compare-genome-synteny" class="anchor"></a>Compare genome synteny</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># All-vs-all alignment | https://github.com/lh3/minimap2</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="ex">minimap2</span> -X -N 50 -p 0.1 -c emales.fna emales.fna <span class="op">&gt;</span> emales.paf</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emale_links</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_paf.html">read_paf</a></span><span class="op">(</span><span class="st">"emales.paf"</span><span class="op">)</span>

<span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">emale_seqs_6</span>, <span class="va">emale_genes</span>, <span class="va">emale_tirs</span>, <span class="va">emale_links</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_feature</span><span class="op">(</span>size<span class="op">=</span><span class="fl">5</span>, data<span class="op">=</span><span class="fu">use_features</span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">gc_cont</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_link.html">geom_link</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_distiller</span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Spectral"</span><span class="op">)</span>

<span class="va">p4</span> <span class="op">&lt;-</span> <span class="va">p4</span> <span class="op">%&gt;%</span> <span class="fu">flip_bins</span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>
<span class="va">p4</span></code></pre></div>
<p><img src="emales/emales-p4.png"></p>
</div>
<div id="gc-content" class="section level3">
<h3 class="hasAnchor">
<a href="#gc-content" class="anchor"></a>GC-content</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># https://github.com/thackl/seq-scripts (bedtools &amp; samtools)</span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="ex">seq-gc</span> -Nbw 50 emales.fna <span class="op">&gt;</span> emales-gc.tsv</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emale_gc</span> <span class="op">&lt;-</span> <span class="fu">thacklr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/thacklr/man/read_bed.html">read_bed</a></span><span class="op">(</span><span class="st">"emales-gc.tsv"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">rename</span><span class="op">(</span>seq_id<span class="op">=</span><span class="va">contig_id</span><span class="op">)</span>

<span class="va">p5</span> <span class="op">&lt;-</span> <span class="va">p4</span> <span class="op">%&gt;%</span> <span class="fu">add_features</span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span>
<span class="va">p5</span> <span class="op">&lt;-</span> <span class="va">p5</span> <span class="op">+</span> <span class="fu">geom_ribbon</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="op">(</span><span class="va">x</span><span class="op">+</span><span class="va">xend</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, ymax<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">.24</span>, ymin<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">.38</span><span class="op">-</span><span class="op">(</span><span class="fl">.4</span><span class="op">*</span><span class="va">score</span><span class="op">)</span>,
    group<span class="op">=</span><span class="va">seq_id</span>, linetype<span class="op">=</span><span class="st">"GC-content"</span><span class="op">)</span>, <span class="fu">use_features</span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span>,
                       fill<span class="op">=</span><span class="st">"blue"</span>, alpha<span class="op">=</span><span class="fl">.5</span><span class="op">)</span>
<span class="va">p5</span></code></pre></div>
<p><img src="emales/emales-p5.png"></p>
</div>
<div id="cluster-protein-sequences-into-orthogroups" class="section level3">
<h3 class="hasAnchor">
<a href="#cluster-protein-sequences-into-orthogroups" class="anchor"></a>cluster protein sequences into orthogroups</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="ex">gff2cds</span> --aa --type CDS --source Prodigal_v2.6.3 --fna emales.fna emales.gff <span class="op">&gt;</span> emales.faa</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="ex">mmseqs</span> easy-cluster emales.faa emales-mmseqs /tmp -e 1e-5 -c 0.7<span class="kw">;</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="ex">cluster-ids</span> -t <span class="st">"cog%03d"</span> <span class="op">&lt;</span> emales-mmseqs_cluster.tsv <span class="op">&gt;</span> emales-cogs.tsv</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emale_cogs</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">"emales-cogs.tsv"</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"feature_id"</span>, <span class="st">"cluster_id"</span>, <span class="st">"cluster_n"</span><span class="op">)</span><span class="op">)</span>
<span class="va">emale_cogs</span> <span class="op">%&lt;&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>
  cluster_label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">cluster_id</span>, <span class="st">" ("</span>, <span class="va">cluster_n</span>, <span class="st">")"</span><span class="op">)</span>,
  cluster_label <span class="op">=</span> <span class="fu">fct_lump_min</span><span class="op">(</span><span class="va">cluster_label</span>, <span class="fl">5</span>, other_level <span class="op">=</span> <span class="st">"rare"</span><span class="op">)</span>,
  cluster_label <span class="op">=</span> <span class="fu">fct_lump_min</span><span class="op">(</span><span class="va">cluster_label</span>, <span class="fl">15</span>, other_level <span class="op">=</span> <span class="st">"medium"</span><span class="op">)</span>,
  cluster_label <span class="op">=</span> <span class="fu">fct_relevel</span><span class="op">(</span><span class="va">cluster_label</span>, <span class="st">"rare"</span>, after<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span><span class="op">)</span>
<span class="va">emale_cogs</span>


<span class="va">p6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">emale_seqs_6</span>, <span class="va">emale_genes</span>, <span class="va">emale_tirs</span>, <span class="va">emale_links</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">add_features</span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/add_feats.html">add_clusters</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">emale_cogs</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">flip_bins</span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_feature</span><span class="op">(</span>size<span class="op">=</span><span class="fl">5</span>, data<span class="op">=</span><span class="fu">use_features</span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">cluster_label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_link.html">geom_link</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_ribbon</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="op">(</span><span class="va">x</span><span class="op">+</span><span class="va">xend</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, ymax<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">.24</span>, ymin<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">.38</span><span class="op">-</span><span class="op">(</span><span class="fl">.4</span><span class="op">*</span><span class="va">score</span><span class="op">)</span>,
    group<span class="op">=</span><span class="va">seq_id</span>, linetype<span class="op">=</span><span class="st">"GC-content"</span><span class="op">)</span>, <span class="fu">use_features</span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span>,
                       fill<span class="op">=</span><span class="st">"blue"</span>, alpha<span class="op">=</span><span class="fl">.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_brewer</span><span class="op">(</span><span class="st">"Conserved genes"</span>, palette<span class="op">=</span><span class="st">"Set3"</span><span class="op">)</span>

<span class="va">p6</span></code></pre></div>
<p><img src="emales/emales-p6.png"></p>
</div>
<div id="blast-hits-and-integrated-transposons" class="section level3">
<h3 class="hasAnchor">
<a href="#blast-hits-and-integrated-transposons" class="anchor"></a>Blast hits and integrated transposons</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># mavirus.faa - published</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="ex">blastp</span> -query emales.faa -subject mavirus.faa -outfmt 7 <span class="op">&gt;</span> emales_mavirus-blastp.tsv</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="fu">perl</span> -ne <span class="st">'if(/&gt;(\S+) gene=(\S+) product=(.+)/){print join("\t", $1, $2, $3), "\n"}'</span> \</span>
<span id="cb12-4"><a href="#cb12-4"></a>  mavirus.faa <span class="op">&gt;</span> mavirus.tsv</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emale_blast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_blast.html">read_blast</a></span><span class="op">(</span><span class="st">"emales_mavirus-blastp.tsv"</span><span class="op">)</span>
<span class="va">emale_blast</span> <span class="op">%&lt;&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">evalue</span> <span class="op">&lt;</span> <span class="fl">1e-3</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span>feature_id<span class="op">=</span><span class="va">qaccver</span>, start<span class="op">=</span><span class="va">qstart</span>, end<span class="op">=</span><span class="va">qend</span>, <span class="va">saccver</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">left_join</span><span class="op">(</span><span class="fu">read_tsv</span><span class="op">(</span><span class="st">"mavirus.tsv"</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"saccver"</span>, <span class="st">"blast_hit"</span>, <span class="st">"blast_desc"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># manual annotations by MFG</span>
<span class="va">emale_transposons</span> <span class="op">&lt;-</span> <span class="fu">read_gff</span><span class="op">(</span><span class="st">"emales-manual.gff"</span>, types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"mobile_element"</span><span class="op">)</span><span class="op">)</span>


<span class="va">p7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">emale_seqs_6</span>, <span class="va">emale_genes</span>, <span class="va">emale_tirs</span>, <span class="va">emale_links</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">add_features</span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/add_feats.html">add_clusters</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">emale_cogs</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">add_features</span><span class="op">(</span><span class="va">emale_transposons</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">add_subfeatures</span><span class="op">(</span><span class="va">genes</span>, <span class="va">emale_blast</span>, transform<span class="op">=</span><span class="st">"aa2nuc"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">flip_bins</span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_feature</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"integrated transposon"</span><span class="op">)</span>,
    <span class="fu">use_features</span><span class="op">(</span><span class="va">emale_transposons</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">7</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_link.html">geom_link</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.2</span><span class="op">)</span>, color<span class="op">=</span><span class="st">"white"</span>, alpha<span class="op">=</span><span class="fl">.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_feature</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"terminal inverted repeat"</span><span class="op">)</span>, <span class="fu">use_features</span><span class="op">(</span><span class="va">features</span><span class="op">)</span>,
    size<span class="op">=</span><span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">cluster_label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_feature</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">blast_desc</span><span class="op">)</span>, <span class="fu">use_features</span><span class="op">(</span><span class="va">emale_blast</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">2</span>,
    position<span class="op">=</span><span class="st">"pile"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_ribbon</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="op">(</span><span class="va">x</span><span class="op">+</span><span class="va">xend</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, ymax<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">.24</span>, ymin<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">.38</span><span class="op">-</span><span class="op">(</span><span class="fl">.4</span><span class="op">*</span><span class="va">score</span><span class="op">)</span>,
    group<span class="op">=</span><span class="va">seq_id</span>, linetype<span class="op">=</span><span class="st">"GC-content"</span><span class="op">)</span>, <span class="fu">use_features</span><span class="op">(</span><span class="va">emale_gc</span><span class="op">)</span>,
                       fill<span class="op">=</span><span class="st">"blue"</span>, alpha<span class="op">=</span><span class="fl">.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_brewer</span><span class="op">(</span><span class="st">"Conserved genes"</span>, palette<span class="op">=</span><span class="st">"Set3"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_color_viridis_d</span><span class="op">(</span><span class="st">"Blast hits &amp; Features"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_linetype</span><span class="op">(</span><span class="st">"Graphs"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Endogenous mavirus-like elements of "</span>,
  <span class="fu"><a href="https://rdrr.io/r/grDevices/plotmath.html">italic</a></span><span class="op">(</span><span class="st">"C. burkhardae"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">p7</span></code></pre></div>
<p><img src="emales/emales-p7.png"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Thomas Hackl, Markus J. Ankenbrand.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
