<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • gggenomes</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="">
<meta property="og:description" content="gggenomes">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gggenomes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.10.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/gggenomes.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/emales.html">From a few sequences to a complex map in minutes (old version)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thackl/gggenomes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/thackl/gggenomes/blob/HEAD/vignettes/gggenomes.Rmd" class="external-link"><code>vignettes/gggenomes.Rmd</code></a></small>
      <div class="hidden name"><code>gggenomes.Rmd</code></div>

    </div>

    
    
<p>q— title: “gggenomes - a multi-track ggplot2 extension for
comparative genomics” author: “Thomas Hackl” date: “2/23/2021” output:
pdf_document —</p>
<p>gggenomes is an extension to ggplot2 tailored to the visualization of
genomics data. It provides the same flexible approach for building up
plots layer by layer, but supports the simultaneous use of multiple
datasets within one plot, thereby intuitively mirroring the relational
nature of most ’omics data products.</p>
<p><strong>A ggplot2 is build around a single, tidy dataset</strong>
with arbitrary variables and observations. x/y-plot coordinates for
things to draw are computed on-the-fly using aesthetic mappings that
link variables to coordinates. This setup is highly flexible and makes
ggplot extremely powerful.</p>
<p><strong>A gggenomes plot is build around multiple data
tracks</strong> that each contain a single dataset. These datasets are
linked via pre-computed global layout that determines x/y-plot
coordinates prior to the actual plot construction. The reason for this
more rigid setup is that genomics data are relational and there is just
no good way to represent all the necessary information in a single tidy
table.</p>
<p>gggenomes knows three types of tracks:</p>
<ol style="list-style-type: decimal">
<li>
<strong>seqs</strong>: sequences such as contigs or chromosomes</li>
<li>
<strong>feats</strong>: annotations of locations on sequences, such
genes, SNPs, …</li>
<li>
<strong>links</strong>: annotations that connect two locations
between two different sequences</li>
</ol>
<p>There are also <strong><em>genes</em></strong> tracks, those, however
are just glorified feature tracks, that sometimes get special treatment
for easier usage. Internally, genes are features.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thackl/gggenomes" class="external-link">gggenomes</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::%@%' by 'rlang::%@%' when loading</span></span>
<span><span class="co">## 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::flatten_lgl' by 'rlang::flatten_lgl'</span></span>
<span><span class="co">## when loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::splice' by 'rlang::splice' when</span></span>
<span><span class="co">## loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::flatten_chr' by 'rlang::flatten_chr'</span></span>
<span><span class="co">## when loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::flatten_raw' by 'rlang::flatten_raw'</span></span>
<span><span class="co">## when loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::flatten' by 'rlang::flatten' when</span></span>
<span><span class="co">## loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::flatten_dbl' by 'rlang::flatten_dbl'</span></span>
<span><span class="co">## when loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::invoke' by 'rlang::invoke' when</span></span>
<span><span class="co">## loading 'gggenomes'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'purrr::flatten_int' by 'rlang::flatten_int'</span></span>
<span><span class="co">## when loading 'gggenomes'</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a minimal seq track</span></span>
<span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fl">550</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># a minimal gene track</span></span>
<span><span class="va">g0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">350</span>, <span class="fl">80</span><span class="op">)</span>,</span>
<span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">250</span>, <span class="fl">500</span>, <span class="fl">450</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># a simple link track</span></span>
<span><span class="va">l0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"a"</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">250</span>, <span class="fl">480</span><span class="op">)</span>,</span>
<span>  seq_id2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"b"</span>, <span class="st">"b"</span><span class="op">)</span>,</span>
<span>  start2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">80</span>, <span class="fl">350</span><span class="op">)</span>,</span>
<span>  end2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">430</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span>genes<span class="op">=</span><span class="va">g0</span>, seqs<span class="op">=</span><span class="va">s0</span>, links<span class="op">=</span><span class="va">l0</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>         <span class="co"># draw contig/chromosome lines</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>   <span class="co"># label each sequence </span></span>
<span>  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>        <span class="co"># draw genes as arrow</span></span>
<span>  <span class="fu"><a href="../reference/geom_link.html">geom_link</a></span><span class="op">(</span><span class="op">)</span>          <span class="co"># draw some connections between syntenic regions</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="geoms-and-tracks">Geoms and tracks<a class="anchor" aria-label="anchor" href="#geoms-and-tracks"></a>
</h2>
<p>In a regular ggplot2, geoms read from one central dataframe. For
gggenomes, there are several tracks to read data from. To specify which
tracks to use in which geoms, gggenomes provides accessory functions
that work inside geom calls when provided to the <code>data</code>
argument (or outside if prefixed with <code>pull_</code>). The names of
these functions mirror the track types they work on and should be quite
intuitive:</p>
<ul>
<li>
<code>geom_seq()  : data=seqs()</code> for the seq track</li>
<li>
<code>geom_feat() : data=feats()</code> for the first feat track not
named “genes”</li>
<li>
<code>geom_link() : data=links()</code> for the first link
track</li>
<li>
<code>geom_gene() : data=genes()</code> for the first feat track
with some extras for geneish features (see <a href="#inside-gggenomes-tracks">Inside gggenomes tracks</a> for
details)</li>
</ul>
<p>gggenomes supports an arbitrary number of feature and link tracks
within the same plot. Extra tracks can be added by either providing a
list of tracks to <code><a href="../reference/gggenomes.html">gggenomes()</a></code>, or by using the dedicated
<code><a href="../reference/add_tracks.html">add_feats()</a></code> / <code><a href="../reference/add_tracks.html">add_links()</a></code> functions. The extra
tracks can then be used via the aforementioned accessory function inside
geom calls using names or positional arguments. Like most gggenomes
functions the track accessory functions use tidyverse-style non-standard
evaluation, which means you can refer to tracks either by unquoted names
or by numeric position.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's use some of the bundled example data here</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"gggenomes"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span></span>
<span>  genes<span class="op">=</span><span class="va">emale_genes</span>,  <span class="co"># a gene track, added as first feat track</span></span>
<span>  seqs<span class="op">=</span><span class="va">emale_seqs</span>,    <span class="co"># a seq track</span></span>
<span>  feats<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">emale_tirs</span>, <span class="va">emale_ngaros</span><span class="op">)</span>,  <span class="co"># multiple feat tracks</span></span>
<span>  links<span class="op">=</span><span class="va">emale_ava</span>     <span class="co"># a link track</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the tracks of the plot</span></span>
<span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">track_info</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 5 × 4</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   type [3]</span></span></span>
<span><span class="co">##   id           type      i     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> seqs         seqs      1     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> genes        feats     1   143</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> emale_tirs   feats     2    12</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> emale_ngaros feats     3     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> links        links     1    38</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot all tracks</span></span>
<span><span class="va">p</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_link.html">geom_link</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># the first link track</span></span>
<span>  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># the first feat track filtered for geneish feats: CDS, mRNA, ..</span></span>
<span>  <span class="fu"><a href="../reference/geom_feat.html">geom_feat</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># the first feat track not named "genes", here emale_tirs</span></span>
<span>  <span class="co"># use an extra feat track by name</span></span>
<span>  <span class="fu"><a href="../reference/geom_feat.html">geom_feat</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="../reference/pull_track.html">feats</a></span><span class="op">(</span><span class="va">emale_ngaros</span><span class="op">)</span>, color<span class="op">=</span><span class="st">"plum3"</span><span class="op">)</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-2-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="one-layout-to-rule-them-all">One layout to rule them all<a class="anchor" aria-label="anchor" href="#one-layout-to-rule-them-all"></a>
</h2>
<p>gggenomes multi-track setup has both pros and cons. For example,
communicating information between tracks during plot construction is not
supported by ggplot2, because - well - it’s been designed to work with a
single table. gggenomes works around that by <strong>pre-computing a
layout</strong>, and adding coordinates (<code>y,x,xend</code>) to each
dataframe prior to the actual plot construction. This has some
implications for the usage of gggenomes:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Dataframes for tracks have required variables</strong>,
e.g. <code>seq_id,start,end</code> for features. These predefined
variables are used during import to compute x/y coordinates (see <a href="#inside-gggenomes-tracks">Inside gggenomes tracks</a> for more
details).</li>
<li>
<strong>gggenomes geoms can often be used without explicit
<code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code> mappings</strong> - a rather convenient upside of the
multi-track setup. This works because we always know the names of the
plot variables ahead of time: they originate from the pre-computed
layout, and we can use that information to set sensible default
aesthetic mappings for most cases.</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inspect seqs track with layout vars - note y,x,xend</span></span>
<span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">pull_seqs</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 12</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   bin_id [6]</span></span></span>
<span><span class="co">##       y     x  xend strand seq_id   bin_id length bin_offset start   end file_id</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     6     0 <span style="text-decoration: underline;">20</span>152 +      RCC970_… RCC97…  <span style="text-decoration: underline;">20</span>152          0     1 <span style="text-decoration: underline;">20</span>152 emales </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     5     0 <span style="text-decoration: underline;">20</span>642 +      E4-10_0… E4-10…  <span style="text-decoration: underline;">20</span>642          0     1 <span style="text-decoration: underline;">20</span>642 emales </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     4     0 <span style="text-decoration: underline;">21</span>311 +      Cflag_0… Cflag…  <span style="text-decoration: underline;">21</span>311          0     1 <span style="text-decoration: underline;">21</span>311 emales </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>     3     0 <span style="text-decoration: underline;">26</span>808 +      BVI_069  BVI_0…  <span style="text-decoration: underline;">26</span>808          0     1 <span style="text-decoration: underline;">26</span>808 emales </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>     2     0 <span style="text-decoration: underline;">26</span>820 +      BVI_008A BVI_0…  <span style="text-decoration: underline;">26</span>820          0     1 <span style="text-decoration: underline;">26</span>820 emales </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>     1     0 <span style="text-decoration: underline;">26</span>856 +      E4-10_1… E4-10…  <span style="text-decoration: underline;">26</span>856          0     1 <span style="text-decoration: underline;">26</span>856 emales </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: seq_desc &lt;chr&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inspect genes track with layout vars - note y,x,xend, but also other</span></span>
<span><span class="co"># columns such as strand, feat_id or type, that are added automatically</span></span>
<span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">pull_genes</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 143 × 22</span></span></span>
<span><span class="co">##        y     x  xend bin_id   seq_id   start   end file_id strand type  feat_id </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>     2   567   822 BVI_008A BVI_008A   568   822 emales  +      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>     2  <span style="text-decoration: underline;">2</span>668  <span style="text-decoration: underline;">1</span>039 BVI_008A BVI_008A  <span style="text-decoration: underline;">1</span>040  <span style="text-decoration: underline;">2</span>668 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>     2  <span style="text-decoration: underline;">3</span>028  <span style="text-decoration: underline;">2</span>767 BVI_008A BVI_008A  <span style="text-decoration: underline;">2</span>768  <span style="text-decoration: underline;">3</span>028 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>     2  <span style="text-decoration: underline;">4</span>776  <span style="text-decoration: underline;">3</span>138 BVI_008A BVI_008A  <span style="text-decoration: underline;">3</span>139  <span style="text-decoration: underline;">4</span>776 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>     2  <span style="text-decoration: underline;">5</span>370  <span style="text-decoration: underline;">4</span>809 BVI_008A BVI_008A  <span style="text-decoration: underline;">4</span>810  <span style="text-decoration: underline;">5</span>370 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>     2  <span style="text-decoration: underline;">6</span>065  <span style="text-decoration: underline;">5</span>396 BVI_008A BVI_008A  <span style="text-decoration: underline;">5</span>397  <span style="text-decoration: underline;">6</span>065 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>     2  <span style="text-decoration: underline;">6</span>683  <span style="text-decoration: underline;">6</span>074 BVI_008A BVI_008A  <span style="text-decoration: underline;">6</span>075  <span style="text-decoration: underline;">6</span>683 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>     2  <span style="text-decoration: underline;">6</span>867  <span style="text-decoration: underline;">6</span>540 BVI_008A BVI_008A  <span style="text-decoration: underline;">6</span>541  <span style="text-decoration: underline;">6</span>867 emales  -      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>     2  <span style="text-decoration: underline;">6</span>954  <span style="text-decoration: underline;">7</span>788 BVI_008A BVI_008A  <span style="text-decoration: underline;">6</span>955  <span style="text-decoration: underline;">7</span>788 emales  +      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>     2  <span style="text-decoration: underline;">7</span>997  <span style="text-decoration: underline;">9</span>026 BVI_008A BVI_008A  <span style="text-decoration: underline;">7</span>998  <span style="text-decoration: underline;">9</span>026 emales  +      CDS   BVI_008…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 133 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 11 more variables: introns &lt;list&gt;, parent_ids &lt;list&gt;, source &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   score &lt;chr&gt;, phase &lt;int&gt;, width &lt;chr&gt;, gc_content &lt;chr&gt;, name &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Note &lt;chr&gt;, geom_id &lt;chr&gt;, .marginal &lt;lgl&gt;</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="plotting-from-scratch">Plotting from scratch<a class="anchor" aria-label="anchor" href="#plotting-from-scratch"></a>
</h2>
<p>The minimum requirement to create a gggenomes plot is a single track.
It can be of any of the three types: seqs, feats or links.</p>
<p>If there is no a sequence track, a sequence track is inferred from
the data simply by assuming that each sequence that a feature or link
maps onto must exist, and that it must be long enough to accommodate all
its associated features and links (note the corresponding log message
below).</p>
<p>Note also, that unless sequence information is explicitly provided,
gggenomes will zoom in on only the regions that are covered with
features. That may appear odd at first, but it is a) consistent in the
sense that for inferred sequence we only know the start
(<code>x=1</code>) anyway, and have to guess the end from the last
feature, and b) it is very convenient if we deal with real data: It
usually makes no sense to plot all 5000 genes of a bacterial genome. In
most cases we will want to focus on specific regions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># some genes</span></span>
<span><span class="va">g0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">350</span>, <span class="fl">80</span><span class="op">)</span>,</span>
<span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">250</span>, <span class="fl">500</span>, <span class="fl">450</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">g0</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## No seqs provided, inferring seqs from feats</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>         <span class="co"># draw contig/chromosome lines</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>   <span class="co"># label each sequence </span></span>
<span>  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span>          <span class="co"># draw genes as arrow</span></span></code></pre></div>
<pre><code><span><span class="co">## Only saw `type=NA` in genes and will treat everything as `type="CDS"`.</span></span></code></pre>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="plotting-from-files">Plotting from files<a class="anchor" aria-label="anchor" href="#plotting-from-files"></a>
</h2>
<p>Plotting from scratch is fun, but likely not the typical use-case.
Usually you will have files with data that you want to import. gggenomes
provides a bunch of <code>read_*</code> functions that help you do that
for several common bioinformatics file formats. Moreover, gggenomes come
with a set of higher-level convenience read functions that perfectly
integrate with its track system.</p>
<p>What I mean by that is that instead of reading a certain file format
explicitly, such as <code>read_gff3("foo.gff")</code>, you can invoke
the generic <code>read_feats("foo.gff")</code> function to read features
from various file formats. File formats are automatically recognized, so
this, for example, also works <code>read_feats("foo.bed")</code>. And
you can read different information from the same file:
<code>read_seqs("foo.gff")</code> will retrieve sequence information
from the gff3 file instead of gene/feature coordinates.</p>
<p>That said, the only real requirement for any data you want to plot is
that it is in a somewhat tidy table, and that it has some of the
required variables (see below). Even if you have data that is not
directly supported by gggenomes read functions, it should be very easy
to convert your data into a table that works with gggenomes.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># note: ex() is just a helper to get stable paths to gggenomes example data</span></span>
<span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_tracks.html">read_seqs</a></span><span class="op">(</span><span class="fu"><a href="../reference/ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.fna"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_tracks.html">read_feats</a></span><span class="op">(</span><span class="fu"><a href="../reference/ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.gff"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">g0</span>, <span class="va">s0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for lazy people</span></span>
<span><span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="fu"><a href="../reference/ex.html">ex</a></span><span class="op">(</span><span class="st">"emales/emales.gff"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># and really fancy: multiple remote files, all at once</span></span>
<span><span class="va">gbk_phages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  PSSP7 <span class="op">=</span> <span class="st">"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/858/745/GCF_000858745.1_ViralProj15134/GCF_000858745.1_ViralProj15134_genomic.gff.gz"</span>,</span>
<span>  PSSP3 <span class="op">=</span> <span class="st">"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/904/555/GCF_000904555.1_ViralProj195517/GCF_000904555.1_ViralProj195517_genomic.gff.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">gbk_phages</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-5-3.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="inside-gggenomes-tracks">Inside gggenomes tracks<a class="anchor" aria-label="anchor" href="#inside-gggenomes-tracks"></a>
</h2>
<p>As mentioned before, tracks have required columns and optional
columns, which can have additional meaning in the context of the
plot.</p>
<p><strong>seqs</strong>: sequences such as contigs or chromosomes</p>
<ul>
<li>required: <code>seq_id,length</code>
</li>
<li>recognized: <code>bin_id,start,end,strand</code>
</li>
</ul>
<p><strong>feats</strong>: such as genes, SNPs, repeats, …</p>
<ul>
<li>required: <code>seq_id,start,end</code>
</li>
<li>recognized: <code>strand,bin_id,feat_id,introns</code>
</li>
</ul>
<p><strong>links</strong>: connections between two locations from
different sequences</p>
<ul>
<li>required: <code>seq_id,seq_id2</code>
</li>
<li>recognized:
<code>start,end,bin_id,start2,end2,bin_id2,strand</code>
</li>
</ul>
<div class="section level3">
<h3 id="sequence-track---there-can-be-only-one">Sequence track - there can be only one<a class="anchor" aria-label="anchor" href="#sequence-track---there-can-be-only-one"></a>
</h3>
<p><strong><code>seq_id,length</code></strong> are the required
variables and should be quite self-explanatory. It’s highly recommended
(and might become compulsory in the future) for <code>seq_id</code>s to
be unique within the entire data set.</p>
<p>The optional column for seqs are:</p>
<p><strong><code>bin_id</code></strong> to denote <strong>bins</strong>
representing genomes or assemblies with multiple chromosomes or contigs.
Sequences from the same bin are drawn next to each other in the default
layout, as opposed to sequence from different bins which will occupy
different rows. If omitted, the default is to set
<code>bin_id=seq_id</code>, which means, every sequence gets its own
row.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># seq track: one entry per sequence</span></span>
<span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  bin_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a1"</span>,<span class="st">"a2"</span>,<span class="st">"b1"</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2e5</span>, <span class="fl">3e5</span>, <span class="fl">5e5</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs<span class="op">=</span><span class="va">s0</span><span class="op">)</span> </span>
<span><span class="va">p</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>         <span class="co"># draw contig/chromosome lines</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>   <span class="co"># label each sequence </span></span>
<span>  <span class="fu"><a href="../reference/geom_bin_label.html">geom_bin_label</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># label each bin</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;"></p>
<p><strong><code>start,end,strand</code></strong> in the context of a
sequence denote a region or <strong><em>locus</em></strong> on that
sequence. If provided, only the part of the sequence within the locus
boundaries will be shown in the plot. This also will remove any features
from the plot that do not fall within the locus. It’s likely rare to one
wants to specify these information directly, but they are important as
they power manipulation functions such as <code><a href="../reference/focus.html">focus()</a></code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># zoom in on a longer sequence - note the scale on the x-axis</span></span>
<span><span class="va">s0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="st">"a1"</span>,</span>
<span>  length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  start <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  end <span class="op">=</span> <span class="fl">3000</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs<span class="op">=</span><span class="va">s0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="feature-tracks---what-maps-onto-sequences">Feature tracks - what maps onto sequences<a class="anchor" aria-label="anchor" href="#feature-tracks---what-maps-onto-sequences"></a>
</h3>
<p>Again, the required variables <code>seq_id,start,end</code> should be
self-explainatory. Note though, that features with <code>seq_id</code>s
not present in the seq track will be silently ignored.</p>
<p>Also note that gggenomes uses <strong>1-based, inclusive
ranges</strong> to denote feature start/end coordinates. It’s the same
system as used by GFF, GenBank, BLAST or SAM, for example. In contrast,
BED and BAM, for example, use 0-based, exclusive ranges. Something to
keep in mind when importing data from those formats.</p>
<p>The optional columns for features are:</p>
<p><strong><code>strand</code></strong> to denote the orientation of a
feature relative to the sequences it maps to. The default way of
specifying the strand is with a character vector with
<code>"+/-/."</code> coding for <em>forward</em>, <em>reverse</em> and
<em>undetermined</em>. Logical (<code>TRUE/FALSE/NA</code>) or numeric
vectors (<code>1/-1/0</code>) are supported as well.</p>
<p><strong><code>bin_id</code></strong> works as for seqs. The
assignment of <code>bin_id</code>s to features is usually only useful,
if no sequences are provided. Otherwise, bin information is inherited
from the sequences.</p>
<p><strong><code>feat_id</code></strong> is a unique ID for each
feature. These IDs are useful if we want to add data to a plot that
pertains to features and not sequences. For example, we can add the
results of a protein blast search to the genes the proteins derived from
if by matching blast results and feature annotations using the
<code>feat_id</code>. If not provided, they will be automatically
generated.</p>
<p>Special variables for complex gene models:</p>
<p><strong><code>introns</code></strong> is a special column that can
contain a list of numeric vectors that denote intron start and end
coordinates within a given feature. The support of introns is something
I just very recently added. It’s still experimental and some of the
handling might change in future versions.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># some genes</span></span>
<span><span class="va">g0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">800</span><span class="op">)</span>,</span>
<span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">1200</span><span class="op">)</span>,</span>
<span>  <span class="co"># NOTE: introns need to be a list-column!</span></span>
<span>  introns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>,<span class="fl">200</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>,<span class="fl">200</span>,<span class="fl">250</span>,<span class="fl">300</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span><span class="va">g0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>         <span class="co"># draw contig/chromosome lines</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>   <span class="co"># label each sequence </span></span>
<span>  <span class="fu"><a href="../reference/geom_gene.html">geom_gene</a></span><span class="op">(</span><span class="op">)</span>          <span class="co"># draw genes as arrow</span></span></code></pre></div>
<pre><code><span><span class="co">## No seqs provided, inferring seqs from feats</span></span>
<span><span class="co">## Only saw `type=NA` in genes and will treat everything as `type="CDS"`.</span></span></code></pre>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="link-tracks---show-me-the-synteny">Link tracks - show me the synteny<a class="anchor" aria-label="anchor" href="#link-tracks---show-me-the-synteny"></a>
</h3>
<p>The link track connects loci/regions on two adjacent
sequences/genomes with one another. <strong>seq_id, seq_id2</strong> are
the required variables for the link track. Once again, links with
<code>seq_id</code>s not present in the sequence (<code>seq</code>)
track will be silently ignored.<br>
To correctly link specific loci/regions between sequences with one
another, it is also highly recommend to use the following optional
variables: <em>(Without the variables, <code>start</code>,
start2<code>,</code>end<code>and</code>end2`, links will be made between
two entire sequences/contigs and not between two specific
loci/regions.)</em></p>
<p><strong><code>start</code>, <code>start2</code></strong> indicates
the starting position of the loci/region that will be linked between
respectively the first sequence (seq_id) and the second sequence
(seq_id2). <em>Note that links can and will only be created between
adjacent sequences.</em></p>
<p><strong><code>end</code>, <code>end2</code></strong> similar to
<code>start</code> and <code>start2</code>, but now indicates the end of
the loci/region that will be linked. Once again respectively for seq_id
and seq_id2.</p>
<p><strong><code>bin_id</code>, <code>bin_id2</code></strong> works as
for seqs. The assignment of <code>bin_id</code>s to features is usually
only useful, if no sequences are provided. Otherwise, bin information is
inherited from the sequences.</p>
<p><strong><code>strand</code></strong>, works as for feats.
<code>strand</code> denotes the orientation of the link relative to each
of the two sequences. If abscent, the orientation is derived from the
start and end positions.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># some links</span></span>
<span><span class="va">l0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"a"</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">801</span>, <span class="fl">1600</span><span class="op">)</span>,</span>
<span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">550</span>, <span class="fl">1300</span>, <span class="fl">1800</span><span class="op">)</span>,</span>
<span>  seq_id2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"b"</span>, <span class="st">"b"</span>, <span class="st">"b"</span><span class="op">)</span>,</span>
<span>  start2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1100</span>, <span class="fl">1</span>, <span class="fl">1800</span><span class="op">)</span>,</span>
<span>  end2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1450</span>, <span class="fl">500</span>, <span class="fl">1600</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># corresponding sequences</span></span>
<span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  seq_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>, </span>
<span>  length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">2000</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/gggenomes.html">gggenomes</a></span><span class="op">(</span>seqs<span class="op">=</span><span class="va">s1</span>, links<span class="op">=</span><span class="va">l0</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/geom_seq.html">geom_seq</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>                  <span class="co"># draws contigs/chromosome lines  </span></span>
<span>  <span class="fu"><a href="../reference/geom_seq_label.html">geom_seq_label</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>            <span class="co"># labels each sequence</span></span>
<span>  <span class="fu"><a href="../reference/geom_link.html">geom_link</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>      <span class="co"># draws links between contigs</span></span></code></pre></div>
<p><img src="gggenomes_files/figure-html/unnamed-chunk-9-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Thomas Hackl, Markus J. Ankenbrand, Bart van Adrichem.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
